--- stunnel-5.55.orig/src/sthreads.c	2019-07-03 14:25:02.421666749 +1000
+++ stunnel-5.55/src/sthreads.c	2019-07-03 14:39:01.499001206 +1000
@@ -102,17 +102,16 @@
 
 #endif /* USE_WIN32 */
 
-#if OPENSSL_VERSION_NUMBER>=0x10000000L && OPENSSL_VERSION_NUMBER<0x10100004L
+#if OPENSSL_VERSION_NUMBER>=0x10000000L
 NOEXPORT void threadid_func(CRYPTO_THREADID *tid) {
     CRYPTO_THREADID_set_numeric(tid, stunnel_thread_id());
 }
 #endif
 
 void thread_id_init(void) {
-#if OPENSSL_VERSION_NUMBER>=0x10000000L && OPENSSL_VERSION_NUMBER<0x10100000L
+#if OPENSSL_VERSION_NUMBER>=0x10000000L
     CRYPTO_THREADID_set_callback(threadid_func);
-#endif
-#if OPENSSL_VERSION_NUMBER<0x10000000L || !defined(OPENSSL_NO_DEPRECATED)
+#elif OPENSSL_VERSION_NUMBER<0x10000000L || !defined(OPENSSL_NO_DEPRECATED)
     CRYPTO_set_id_callback(stunnel_thread_id);
 #endif
 }
