--- lib/axtls/crypto/crypto_misc.c.org	2020-10-07 14:31:40.283385884 +1000
+++ lib/axtls/crypto/crypto_misc.c	2020-10-07 14:38:47.147774710 +1000
@@ -122,8 +122,9 @@
     }
 #elif !defined(__ets__)
     /* start of with a stack to copy across */
-    int i;
-    memcpy(entropy_pool, &i, ENTROPY_POOL_SIZE);
+    int i = ENTROPY_POOL_SIZE;
+    void *ip = (void *) &i;
+    memcpy(entropy_pool, ip, ENTROPY_POOL_SIZE);
     rand_r((unsigned int *)entropy_pool); 
 #endif
 }
