--- lib/axtls/ssl/tls1.c	2017-12-06 21:17:03.273185147 +1000
+++ lib/axtls/ssl/tls1.c	2017-12-06 21:17:05.877185204 +1000
@@ -2012,11 +2012,15 @@
     int cert_size, offset = 5, offset_start;
     int total_cert_len = (buf[offset]<<8) + buf[offset+1];
     int is_client = IS_SET_SSL_FLAG(SSL_IS_CLIENT);
+#ifndef CONFIG_SSL_SKELETON_MODE
     X509_CTX *chain = 0;
+#endif
     X509_CTX **certs = 0;
     int *cert_used = 0;
     int num_certs = 0;
+#ifndef CONFIG_SSL_SKELETON_MODE
     int i = 0;
+#endif
     offset += 2;
 
     PARANOIA_CHECK(pkt_size, total_cert_len + offset);
@@ -2067,8 +2071,10 @@
     PARANOIA_CHECK(pkt_size, offset);
 
     // third pass - link certs together, assume server cert is the first
+#ifndef CONFIG_SSL_SKELETON_MODE
     *x509_ctx = certs[0];
     chain = certs[0];
+#endif
     cert_used[0] = 1;
 
 #ifndef CONFIG_SSL_SKELETON_MODE
