--- lib/axtls/ssl/tls1_svr.c.org	2020-10-07 14:47:47.745837459 +1000
+++ lib/axtls/ssl/tls1_svr.c	2020-10-07 14:49:36.491831452 +1000
@@ -36,11 +36,6 @@
 #ifdef CONFIG_SSL_ENABLE_SERVER
 
 static const uint8_t g_hello_done[] = { HS_SERVER_HELLO_DONE, 0, 0, 0 };
-static const uint8_t g_asn1_sha256[] = 
-{ 
-    0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 
-    0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20
-};
 
 static int process_client_hello(SSL *ssl);
 static int send_server_hello_sequence(SSL *ssl);
@@ -435,6 +430,12 @@
 
 static const uint8_t g_cert_request_v1[] = { HS_CERT_REQ, 0, 0, 4, 1, 0, 0, 0 };
 
+static const uint8_t g_asn1_sha256[] = 
+{ 
+    0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 
+    0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20
+};
+
 /*
  * Send the certificate request message.
  */
