--- lib/axtls/ssl/asn1.c.org	2020-10-07 14:43:12.906623321 +1000
+++ lib/axtls/ssl/asn1.c	2020-10-07 14:44:58.904504172 +1000
@@ -75,21 +75,6 @@
     0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x03
 };
 
-static const uint8_t sig_subject_alt_name[] =
-{
-    0x55, 0x1d, 0x11
-};
-
-static const uint8_t sig_basic_constraints[] =
-{
-    0x55, 0x1d, 0x13
-};
-
-static const uint8_t sig_key_usage[] =
-{
-    0x55, 0x1d, 0x0f
-};
-
 /* CN, O, OU, L, C, ST */
 static const uint8_t g_dn_types[] = { 3, 10, 11, 7, 6, 8 };
 
@@ -556,6 +541,21 @@
 }
 
 #ifdef CONFIG_SSL_CERT_VERIFICATION
+static const uint8_t sig_subject_alt_name[] =
+{
+    0x55, 0x1d, 0x11
+};
+
+static const uint8_t sig_basic_constraints[] =
+{
+    0x55, 0x1d, 0x13
+};
+
+static const uint8_t sig_key_usage[] =
+{
+    0x55, 0x1d, 0x0f
+};
+
 /**
  * Read the signature of the certificate.
  */
