--- a/disk-utils/Makemodule.am	2023-08-29 03:02:50.137649422 +1000
+++ b/disk-utils/Makemodule.am	2023-08-29 03:04:08.659442471 +1000
@@ -160,6 +160,7 @@
 	disk-utils/fdisk-list.h
 
 fdisk_LDADD = $(LDADD) libcommon.la libfdisk.la \
+	      libuuid.la libblkid.la \
 	      libsmartcols.la libtcolors.la $(READLINE_LIBS)
 fdisk_CFLAGS = $(AM_CFLAGS) -I$(ul_libfdisk_incdir) -I$(ul_libsmartcols_incdir)
 
@@ -183,6 +184,7 @@
 	disk-utils/fdisk-list.h
 
 sfdisk_LDADD = $(LDADD) libcommon.la libfdisk.la \
+	       libuuid.la libblkid.la \
 	       libsmartcols.la libtcolors.la $(READLINE_LIBS)
 sfdisk_CFLAGS = $(AM_CFLAGS) -I$(ul_libfdisk_incdir) -I$(ul_libsmartcols_incdir)
 
@@ -206,6 +208,7 @@
 	libsmartcols.la \
 	libcommon.la \
 	libfdisk.la \
+	libuuid.la libblkid.la \
 	libtcolors.la
 cfdisk_CFLAGS = \
 	$(AM_CFLAGS) \
--- a/sys-utils/Makemodule.am	2023-08-29 04:21:11.872084452 +1000
+++ b/sys-utils/Makemodule.am	2023-08-29 04:22:46.653495713 +1000
@@ -141,7 +141,8 @@
 MANPAGES += sys-utils/fstrim.8
 dist_noinst_DATA += sys-utils/fstrim.8.adoc
 fstrim_SOURCES = sys-utils/fstrim.c
 fstrim_LDADD = $(LDADD) libcommon.la libmount.la
+fstrim_LDADD += libblkid.la
 fstrim_CFLAGS = $(AM_CFLAGS) -I$(ul_libmount_incdir)
 if HAVE_SYSTEMD
 systemdsystemunit_DATA += \
@@ -283,7 +283,8 @@
 MANPAGES += sys-utils/eject.1
 dist_noinst_DATA += sys-utils/eject.1.adoc
 eject_SOURCES =  sys-utils/eject.c lib/monotonic.c
 eject_LDADD = $(LDADD) libmount.la libcommon.la $(REALTIME_LIBS)
+eject_LDADD += libblkid.la
 eject_CFLAGS = $(AM_CFLAGS) -I$(ul_libmount_incdir)
 endif
 
@@ -332,7 +332,8 @@
 MANPAGES += sys-utils/lsns.8
 dist_noinst_DATA += sys-utils/lsns.8.adoc
 lsns_SOURCES =	sys-utils/lsns.c
 lsns_LDADD = $(LDADD) libcommon.la libsmartcols.la libmount.la
+lsns_LDADD += libblkid.la
 lsns_CFLAGS = $(AM_CFLAGS) -I$(ul_libsmartcols_incdir) -I$(ul_libmount_incdir)
 endif
 
@@ -348,12 +348,14 @@
 	sys-utils/fstab.5.adoc \
 	sys-utils/umount.8.adoc
 mount_SOURCES = sys-utils/mount.c
 mount_LDADD = $(LDADD) libcommon.la libmount.la $(SELINUX_LIBS)
+mount_LDADD += libblkid.la
 mount_CFLAGS = $(SUID_CFLAGS) $(AM_CFLAGS) -I$(ul_libmount_incdir)
 mount_LDFLAGS = $(SUID_LDFLAGS) $(AM_LDFLAGS)
 
 umount_SOURCES = sys-utils/umount.c
 umount_LDADD = $(LDADD) libcommon.la libmount.la
+umount_LDADD += libblkid.la
 umount_CFLAGS = $(AM_CFLAGS) $(SUID_CFLAGS) -I$(ul_libmount_incdir)
 umount_LDFLAGS = $(SUID_LDFLAGS) $(AM_LDFLAGS)
 
--- a/misc-utils/Makemodule.am	2023-08-29 04:26:36.539216904 +1000
+++ b/misc-utils/Makemodule.am	2023-08-29 04:28:53.411575456 +1000
@@ -77,6 +77,7 @@
 MANPAGES += misc-utils/lslocks.8
 dist_noinst_DATA += misc-utils/lslocks.8.adoc
 lslocks_LDADD = $(LDADD) libcommon.la libmount.la libsmartcols.la
+lslocks_LDADD += libblkid.la
 lslocks_SOURCES = misc-utils/lslocks.c
 lslocks_CFLAGS = $(AM_CFLAGS) -I$(ul_libmount_incdir) -I$(ul_libsmartcols_incdir)
 endif
