--- asterisk-11.3.0/res/pjproject/build/cc-gcc.mak.orig	2013-05-08 17:03:55.240424349 +1000
+++ asterisk-11.3.0/res/pjproject/build/cc-gcc.mak	2013-05-08 17:04:46.740439025 +1000
@@ -1,8 +1,8 @@
-export CC = $(CROSS_COMPILE)gcc -c
-export AR = $(CROSS_COMPILE)ar rv 
-export LD = $(CROSS_COMPILE)gcc
+export LD := $(CC)
+export CC := $(CC) -c
+export AR := $(AR) rv 
 export LDOUT = -o 
-export RANLIB = $(CROSS_COMPILE)ranlib
+#export RANLIB = $(CROSS_COMPILE)ranlib
 
 export OBJEXT := .o
 export LIBEXT := .a
