--- asterisk-11.3.0/res/Makefile.orig	2013-05-08 13:34:22.891035649 +1000
+++ asterisk-11.3.0/res/Makefile	2013-05-08 13:34:25.390934005 +1000
@@ -71,7 +71,7 @@
 	rm -f snmp/*.o snmp/*.i ael/*.o ael/*.i ais/*.o ais/*.i
 
 pjproject/build.mak:
-	cd pjproject && ./configure AR="" CFLAGS=-fPIC  --disable-floating-point --disable-sound --disable-oss --disable-speex-aec --disable-l16-codec --disable-gsm-codec --disable-g722-codec --disable-g7221-codec --disable-speex-codec --disable-ilbc-codec --disable-g711-codec
+	cd pjproject && ./configure $(CONFIGURE_OPTS) AR="$(AR)" CFLAGS="$(CFLAGS) -fPIC"  --disable-floating-point --disable-sound --disable-oss --disable-speex-aec --disable-l16-codec --disable-gsm-codec --disable-g722-codec --disable-g7221-codec --disable-speex-codec --disable-ilbc-codec --disable-g711-codec
 
 .PHONY : pjproject
 pjproject: pjproject/build.mak
