From 73af7ea579d5580bed8ee9934302d7fbf08cf462 Mon Sep 17 00:00:00 2001
From: Kurt Korbatits <kurt.korbatits@digi.com>
Date: Tue, 8 Nov 2022 13:52:36 +1000
Subject: [PATCH] Allow syslogd to do log rotation

---
 src/syslogd.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/syslogd.c b/src/syslogd.c
index a5edab5..566dc3e 100644
--- a/src/syslogd.c
+++ b/src/syslogd.c
@@ -1498,6 +1498,12 @@ void logrotate(struct filed *f)
 
 	/* bug (mostly harmless): can wrap around if file > 4gb */
 	if (S_ISREG(statf.st_mode) && statf.st_size > f->f_rotatesz) {
+
+		char command[strlen(f->f_un.f_fname) + 10 + 5];
+
+		sprintf(command, "chattr -a %s*", f->f_un.f_fname);
+		system(command);
+
 		if (f->f_rotatecount > 0) { /* always 0..999 */
 			int  len = strlen(f->f_un.f_fname) + 10 + 5;
 			int  i;
@@ -1540,6 +1546,9 @@ void logrotate(struct filed *f)
 			}
 		}
 		ftruncate(f->f_file, 0);
+
+		sprintf(command, "chattr +a %s*", f->f_un.f_fname);
+		system(command);
 	}
 }
 
-- 
2.25.1

