diff --git a/gpsd.hotplug b/gpsd.hotplug
index 9f622cd1..0d4e4191 100644
--- a/gpsd.hotplug
+++ b/gpsd.hotplug
@@ -39,10 +39,40 @@ if [ -n "$GPSD_SOCKET" ]; then
   export GPSD_SOCKET
 fi

-if [ -n "$USBAUTO" ]; then
-  [ "$USBAUTO" = "true" ] || exit 0
+eval $(config start)
+config load
+LOCATION_ENABLE="$(config get service.location.enable)"
+USB_ENABLE=0
+SOURCE_NUM="$(config count service.location.source)"
+for idx in $(seq 0 $((SOURCE_NUM - 1))); do
+	SOURCE_TYPE="$(config get service.location.source.$idx.type)"
+	SOURCE_ENABLED="[ $(config get service.location.source.$idx.enable) -eq 1 ]"
+	if [ "$SOURCE_TYPE" = "usb" ] && $SOURCE_ENABLED; then
+		USB_ENABLE=1
+	fi
+done
+USB_PORT_PATHS=
+for port in $(config keys device.usb.port); do
+	for path in $(config get device.usb.port.${port}.path); do
+		USB_PORT_PATHS="${USB_PORT_PATHS} ${path}"
+	done
+done
+eval $(config stop)
+if [ "$LOCATION_ENABLE" -eq 0 ] || [ "$USB_ENABLE" -eq 0 ]; then
+  exit 0
 fi
 
+# Check if the device is on an external USB port
+path_valid=
+for path in ${USB_PORT_PATHS}; do
+	case ${ID_USB_DEVICE_PATH} in
+	${path}|${path}.*)
+		path_valid=1
+		;;
+	esac
+done
+[ -n "${path_valid}" ] || exit 0
+
 if [ "$ACTION" = "remove" ] ; then
   if echo $DEVLINKS | grep -q /dev/gps; then
     :
