diff --git a/SConstruct b/SConstruct
index b6cc327..0913c9e 100644
--- a/SConstruct
+++ b/SConstruct
@@ -1596,7 +1596,7 @@ if 'dev' in gpsd_version or not os.path.exists('leapseconds.cache'):
 if env['systemd']:
     udevcommand = 'TAG+="systemd", ENV{SYSTEMD_WANTS}="gpsdctl@%k.service"'
 else:
-    udevcommand = 'RUN+="%s/gpsd.hotplug"' % (env['udevdir'], )
+    udevcommand = 'RUN+="%s/gpsd.hotplug"' % ("/libexec/udev", )
 
 
 # Instantiate some file templates.  We'd like to use the Substfile builtin
