--- squid-3.5.21/src/Makefile.in.org	2016-10-06 10:44:28.354396320 +1000
+++ squid-3.5.21/src/Makefile.in	2016-10-06 10:45:42.090394612 +1000
@@ -2955,7 +2955,8 @@
 	$(DISKIO_GEN_SOURCE) \
 	$(BUILT_SOURCES)
 
-squid_LDADD = $(AUTH_ACL_LIBS) ident/libident.la acl/libacls.la \
+squid_LDADD = $(AUTH_ACL_LIBS) $(DISK_LINKOBJS) \
+	ident/libident.la acl/libacls.la \
 	acl/libstate.la $(AUTH_LIBS) $(DISK_LIBS) acl/libapi.la \
 	base/libbase.la libsquid.la ip/libip.la fs/libfs.la \
 	$(SSL_LIBS) ipc/libipc.la mgr/libmgr.la anyp/libanyp.la \
@@ -2963,7 +2964,8 @@
 	http/libsquid-http.la icmp/libicmp.la icmp/libicmp-core.la \
 	log/liblog.la format/libformat.la clients/libclients.la \
 	servers/libservers.la ftp/libftp.la $(XTRA_OBJS) \
-	$(DISK_LINKOBJS) $(REPL_OBJS) $(DISK_OS_LIBS) $(NETTLELIB) \
+	fs/libufs.la fs/librock.la \
+	$(REPL_OBJS) $(DISK_OS_LIBS) $(NETTLELIB) \
 	$(CRYPTLIB) $(REGEXLIB) $(ADAPTATION_LIBS) $(ESI_LIBS) \
 	$(SNMP_LIBS) parser/libsquid-parser.la \
 	$(top_builddir)/lib/libmisccontainers.la \
