--- pam_tacplus/support.c	2021-10-21 13:31:16.808808326 +1000
+++ pam_tacplus.orig/support.c	2021-10-22 11:35:44.223457758 +1000
@@ -278,7 +278,7 @@
         _pam_log(LOG_DEBUG, "%d servers defined", tac_srv_no);
 
         for(n = 0; n < tac_srv_no; n++) {
-            _pam_log(LOG_DEBUG, "server[%d] { addr=%s, key='%s' }", n, tac_ntop(tac_srv[n].addr->ai_addr), tac_srv[n].key);
+            _pam_log(LOG_DEBUG, "server[%d] { addr=%s }", n, tac_ntop(tac_srv[n].addr->ai_addr));
         }
 
         _pam_log(LOG_DEBUG, "tac_service='%s'", tac_service);
