--- pam_tacplus-1.7.0/Makefile.in.orig	2022-11-08 06:17:50.000000000 +1000
+++ pam_tacplus-1.7.0/Makefile.in	2024-07-10 17:50:09.650025347 +1000
@@ -1510,9 +1510,9 @@
 AUTOMAKE_OPTIONS = subdir-objects
 AM_CFLAGS = -Wall -Wextra -Werror
 tacc_SOURCES = tacc.c
-tacc_LDADD = libtac.la $(LTLIBINTL) $(LIB_CRYPTO)  $(LIB_GETRANDOM)
+tacc_LDADD = -L.libs -ltac lib/libgnu.la $(LTLIBINTL) $(LIB_CRYPTO)  $(LIB_GETRANDOM)
 tacc_CFLAGS = $(AM_CFLAGS) -I $(top_srcdir)/libtac/include @rt_debug_defines@ -I $(top_srcdir)/lib
-dist_man_MANS = tacc.1
+#dist_man_MANS = tacc.1
 man1_MANS = tacc.1
 lib_LTLIBRARIES = libtac.la
 libtac_la_includedir = $(includedir)/libtac
@@ -1553,7 +1553,7 @@
 
 pam_tacplus_la_CFLAGS = $(AM_CFLAGS) -I $(top_srcdir)/libtac/include -I $(top_srcdir)/lib
 pam_tacplus_la_LDFLAGS = -module -avoid-version
-pam_tacplus_la_LIBADD = libtac.la
+pam_tacplus_la_LIBADD = -L.libs -ltac
 EXTRA_DIST = pam_tacplus.spec libtac.pc.in m4/gnulib-cache.m4 $(TESTS)
 dist_doc_DATA = sample.pam README.md AUTHORS ChangeLog
 MAINTAINERCLEANFILES = Makefile.in config.h.in configure aclocal.m4 \
