--- lxc/src/lxc/cgroups/cgroup2_devices.c.orig	2021-06-28 13:26:30.858664499 +1000
+++ lxc/src/lxc/cgroups/cgroup2_devices.c	2021-06-28 13:24:50.782666546 +1000
@@ -329,8 +329,9 @@
 
 	prog->kernel_fd = bpf(BPF_PROG_LOAD, attr, sizeof(*attr));
 	if (prog->kernel_fd < 0)
-		return log_error_errno(-1, errno, "Failed to load bpf program: %s",
-				       log_buf ?: "(null)");
+		return errno;
+		/* return log_error_errno(-1, errno, "Failed to load bpf program: %s",
+				       log_buf ?: "(null)"); */
 
 	TRACE("Loaded bpf program: %s", log_buf ?: "(null)");
 	return 0;
