From 2d406e4a407e0bfa7fd7fb5c766d47f662c9a0c6 Mon Sep 17 00:00:00 2001
From: Robert Hodaszi <robert.hodaszi@digi.com>
Date: Fri, 14 Jan 2022 14:20:52 +0100
Subject: [PATCH] Add ucfront compatibility

CC with ucfront is 'ucfront-gcc <gcc>'. Have to split it up for popen().
---
 mesonbuild/compilers/detect.py | 3 +++
 1 file changed, 3 insertions(+)

Index: meson-1.4.1/mesonbuild/compilers/detect.py
===================================================================
--- meson-1.4.1.orig/mesonbuild/compilers/detect.py
+++ meson-1.4.1/mesonbuild/compilers/detect.py
@@ -274,7 +274,9 @@ def _detect_c_or_cpp_compiler(env: 'Envi
 
     for compiler in compilers:
         if isinstance(compiler, str):
             compiler = [compiler]
+        if ' ' in compiler[0]:
+            compiler[:1] = compiler[0].strip().split(' ')
         compiler_name = os.path.basename(compiler[0])
 
         if any(os.path.basename(x) in {'cl', 'cl.exe', 'clang-cl', 'clang-cl.exe'} for x in compiler):
