From 2d406e4a407e0bfa7fd7fb5c766d47f662c9a0c6 Mon Sep 17 00:00:00 2001
From: Robert Hodaszi <robert.hodaszi@digi.com>
Date: Fri, 14 Jan 2022 14:20:52 +0100
Subject: [PATCH] Add ucfront compatibility

CC with ucfront is 'ucfront-gcc <gcc>'. Have to split it up for popen().
---
 mesonbuild/compilers/detect.py | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/mesonbuild/compilers/detect.py b/mesonbuild/compilers/detect.py
index 1eacecf..3f2cf0b 100644
--- a/mesonbuild/compilers/detect.py
+++ b/mesonbuild/compilers/detect.py
@@ -402,6 +402,9 @@ def _detect_c_or_cpp_compiler(env: 'Environment', lang: str, for_machine: Machin
         else:
             arg = '--version'
 
+        if ' ' in compiler[0]:
+            compiler = compiler[0].split()
+
         try:
             p, out, err = Popen_safe(compiler + [arg])
         except OSError as e:
-- 
2.27.0

