--- uClibc-ng-1.0.12/include/signal.h.orig	2016-02-02 01:40:56.000000000 +1000
+++ uClibc-ng-1.0.12/include/signal.h	2016-03-13 22:35:18.032181553 +1000
@@ -214,7 +214,7 @@
 #endif
 #endif /* __UCLIBC_SUSV4_LEGACY__ */
 
-#if 0 /*def __USE_BSD*/
+#ifdef __USE_BSD
 /* None of the following functions should be used anymore.  They are here
    only for compatibility.  A single word (`int') is not guaranteed to be
    enough to hold a complete signal mask and therefore these functions
@@ -222,7 +222,9 @@
 
 /* Compute mask for signal SIG.  */
 # define sigmask(sig)	__sigmask(sig)
+#endif
 
+#if 0
 /* Block signals in MASK, returning the old mask.  */
 extern int sigblock (int __mask) __THROW __attribute_deprecated__;
 
