--- a/lib/audit_logging/audit_logging.c		2021-11-16 00:23:04.560985897 +1000
+++ b/lib/audit_logging/audit_logging.c		2021-11-16 00:26:18.098962996 +1000
@@ -68,9 +68,8 @@
 		return NULL;
 	}
 
-	strftime(buffer, sizeof(buffer)-1, "%a, %d %b %Y %H:%M:%S", tm_info);
-	strftime(tz, sizeof(tz)-1, "%Z", tm_info);
-	ts = talloc_asprintf(frame, "%s.%06ld %s", buffer, (long)tv.tv_usec, tz);
+	strftime(buffer, sizeof(buffer)-1, "%b %d %H:%M:%S", tm_info);
+	ts = talloc_asprintf(frame, "%s", buffer);
 	if (ts == NULL) {
 		DBG_ERR("Out of memory formatting time stamp\n");
 	}
