--- iptables.orig-1.8.7/include/linux/netfilter/xt_mac.h	2021-08-30 16:45:15.176745910 +1000
+++ iptables-1.8.7/include/linux/netfilter/xt_mac.h	2021-08-30 16:45:50.368744867 +1000
@@ -3,6 +3,7 @@
 
 struct xt_mac_info {
     unsigned char srcaddr[ETH_ALEN];
+    unsigned char srcmask[ETH_ALEN];
     int invert;
 };
 #endif /*_XT_MAC_H*/
--- iptables-1.8.7/extensions/libxt_mac.c.orig	2021-01-16 08:03:39.000000000 +1000
+++ iptables-1.8.7/extensions/libxt_mac.c	2021-08-31 13:35:13.354524473 +1000
@@ -9,21 +9,25 @@
 
 enum {
 	O_MAC = 0,
+	O_MASK = 1,
 };
 
 static void mac_help(void)
 {
 	printf(
 "mac match options:\n"
-"[!] --mac-source XX:XX:XX:XX:XX:XX\n"
-"				Match source MAC address\n");
+"[!] --mac-source XX:XX:XX:XX:XX:XX\tMatch source MAC address\n"
+"    --mask XX:XX:XX:XX:XX:XX\tsource address mask\n");
 }
 
 #define s struct xt_mac_info
 static const struct xt_option_entry mac_opts[] = {
 	{.name = "mac-source", .id = O_MAC, .type = XTTYPE_ETHERMAC,
	 .flags = XTOPT_MAND | XTOPT_INVERT | XTOPT_PUT,
 	 XTOPT_POINTER(s, srcaddr)},
+	{.name = "mask", .id = O_MASK, .type = XTTYPE_ETHERMAC,
+	 .flags = XTOPT_PUT,
+	 XTOPT_POINTER(s, srcmask)},
 	XTOPT_TABLEEND,
 };
 #undef s
@@ -48,6 +52,10 @@
 		printf(" !");
 
 	xtables_print_mac(info->srcaddr);
+
+	printf(" MASK");
+
+	xtables_print_mac(info->srcmask);
 }
 
 static void mac_save(const void *ip, const struct xt_entry_match *match)
@@ -59,13 +67,15 @@
 
 	printf(" --mac-source ");
 	xtables_print_mac(info->srcaddr);
+	printf(" --mask ");
+	xtables_print_mac(info->srcmask);
 }
 
 static void print_mac_xlate(const unsigned char *macaddress,
 			    struct xt_xlate *xl)
 {
 	unsigned int i;

 	xt_xlate_add(xl, "%02x", macaddress[0]);
 	for (i = 1; i < ETH_ALEN; ++i)
 		xt_xlate_add(xl, ":%02x", macaddress[i]);
