diff -Nurd iptables.orig-1.8.3/configure.ac iptables-1.8.3/configure.ac
--- iptables.orig-1.8.3/configure.ac	2019-05-30 11:58:09.389798031 +1000
+++ iptables-1.8.3/configure.ac	2019-05-30 11:58:09.465798608 +1000
@@ -54,6 +54,9 @@
 AC_ARG_ENABLE([libipq],
 	AS_HELP_STRING([--enable-libipq], [Build and install libipq]),
 	[enable_libipq="$enableval"], [enable_libipq="no"])
+AC_ARG_ENABLE([osf],
+	AS_HELP_STRING([--enable-osf], [Build OSF util]),
+	[enable_osf="yes"], [enable_osf="no"])
 AC_ARG_ENABLE([bpf-compiler],
 	AS_HELP_STRING([--enable-bpf-compiler], [Build bpf compiler]),
 	[enable_bpfc="$enableval"], [enable_bpfc="no"])
@@ -112,6 +115,7 @@
 AM_CONDITIONAL([ENABLE_LARGEFILE], [test "$enable_largefile" = "yes"])
 AM_CONDITIONAL([ENABLE_DEVEL], [test "$enable_devel" = "yes"])
 AM_CONDITIONAL([ENABLE_LIBIPQ], [test "$enable_libipq" = "yes"])
+AM_CONDITIONAL([ENABLE_OSF], [test "$enable_osf" = "yes"])
 AM_CONDITIONAL([ENABLE_BPFC], [test "$enable_bpfc" = "yes"])
 AM_CONDITIONAL([ENABLE_SYNCONF], [test "$enable_nfsynproxy" = "yes"])
 AM_CONDITIONAL([ENABLE_NFTABLES], [test "$enable_nftables" = "yes"])
@@ -261,6 +265,7 @@
   Devel support:			${enable_devel}
   IPQ support:				${enable_libipq}
   Large file support:			${enable_largefile}
+  OSF util support:			${enable_osf}
   BPF utils support:			${enable_bpfc}
   nfsynproxy util support:		${enable_nfsynproxy}
   nftables support:			${enable_nftables}
diff -Nurd iptables.orig-1.8.3/utils/Makefile.am iptables-1.8.3/utils/Makefile.am
--- iptables.orig-1.8.3/utils/Makefile.am	2019-05-28 01:09:32.000000000 +1000
+++ iptables-1.8.3/utils/Makefile.am	2019-05-30 11:58:09.465798608 +1000
@@ -9,7 +9,7 @@
 dist_pkgdata_DATA =
 man_MANS =
 
-if HAVE_LIBNFNETLINK
+if ENABLE_OSF
 man_MANS += nfnl_osf.8
 sbin_PROGRAMS += nfnl_osf
 dist_pkgdata_DATA += pf.os
