diff -Nurd iptables.orig-1.8.3/libiptc/Makefile.am iptables-1.8.3/libiptc/Makefile.am
--- iptables.orig-1.8.3/libiptc/Makefile.am	2019-05-28 01:09:32.000000000 +1000
+++ iptables-1.8.3/libiptc/Makefile.am	2019-05-30 11:58:09.437798396 +1000
@@ -3,12 +3,9 @@
 AM_CFLAGS        = ${regular_CFLAGS}
 AM_CPPFLAGS      = ${regular_CPPFLAGS} -I${top_builddir}/include -I${top_srcdir}/include ${kinclude_CPPFLAGS}
 
-pkgconfig_DATA      = libiptc.pc libip4tc.pc libip6tc.pc
+pkgconfig_DATA      = libip4tc.pc libip6tc.pc
 
-lib_LTLIBRARIES     = libip4tc.la libip6tc.la libiptc.la
-libiptc_la_SOURCES  =
-libiptc_la_LIBADD   = libip4tc.la libip6tc.la
-libiptc_la_LDFLAGS  = -version-info 0:0:0 ${libiptc_LDFLAGS2}
+lib_LTLIBRARIES     = libip4tc.la libip6tc.la
 libip4tc_la_SOURCES = libip4tc.c
 libip4tc_la_LDFLAGS = -version-info 2:0:0
 libip6tc_la_SOURCES = libip6tc.c
