Index: iptables-1.8.10/configure.ac
===================================================================
--- iptables-1.8.10.orig/configure.ac
+++ iptables-1.8.10/configure.ac
@@ -38,6 +38,9 @@ AC_ARG_ENABLE([ipv4],
 AC_ARG_ENABLE([ipv6],
 	AS_HELP_STRING([--disable-ipv6], [Do not build ip6tables]),
 	[enable_ipv6="$enableval"], [enable_ipv6="yes"])
+AC_ARG_ENABLE([xml],
+	AS_HELP_STRING([--disable-xml], [Do not build iptables-xml]),
+	[enable_xml="$enableval"], [enable_xml="yes"])
 AC_ARG_ENABLE([largefile],
 	AS_HELP_STRING([--disable-largefile], [Do not build largefile support]),
 	[enable_largefile="$enableval"],
@@ -105,6 +108,7 @@ AM_CONDITIONAL([ENABLE_STATIC], [test "$
 AM_CONDITIONAL([ENABLE_SHARED], [test "$enable_shared" = "yes"])
 AM_CONDITIONAL([ENABLE_IPV4], [test "$enable_ipv4" = "yes"])
 AM_CONDITIONAL([ENABLE_IPV6], [test "$enable_ipv6" = "yes"])
+AM_CONDITIONAL([ENABLE_XML], [test "$enable_xml" = "yes"])
 AM_CONDITIONAL([ENABLE_LARGEFILE], [test "$enable_largefile" = "yes"])
 AM_CONDITIONAL([ENABLE_DEVEL], [test "$enable_devel" = "yes"])
 AM_CONDITIONAL([ENABLE_LIBIPQ], [test "$enable_libipq" = "yes"])
@@ -253,6 +257,7 @@ echo "
 Iptables Configuration:
   IPv4 support:				${enable_ipv4}
   IPv6 support:				${enable_ipv6}
+  XML support:				${enable_xml}
   Devel support:			${enable_devel}
   IPQ support:				${enable_libipq}
   Large file support:			${enable_largefile}
Index: iptables-1.8.10/iptables/Makefile.am
===================================================================
--- iptables-1.8.10.orig/iptables/Makefile.am
+++ iptables-1.8.10/iptables/Makefile.am
@@ -6,12 +6,15 @@ AM_LDFLAGS       = ${regular_LDFLAGS}
 
 BUILT_SOURCES =
 
-common_sources = iptables-xml.c xtables-multi.h xshared.c xshared.h
+common_sources = xtables-multi.h xshared.c xshared.h
 common_ldadd   = ../extensions/libext.a ../libxtables/libxtables.la -lm
 common_cflags  = ${AM_CFLAGS}
 if ENABLE_STATIC
 common_cflags += -DALL_INCLUSIVE
 endif
+if ENABLE_XML
+common_sources += iptables-xml.c
+endif
 
 xtables_legacy_multi_SOURCES  = ${common_sources} xtables-legacy-multi.c \
 				iptables-restore.c iptables-save.c
@@ -80,7 +83,9 @@ dist_man_MANS	 = xtables-nft.8 xtables-t
 endif
 CLEANFILES       = ${man_MANS} iptables-extensions.8.tmpl
 
+if ENABLE_XML
 vx_bin_links   = iptables-xml
+endif
 if ENABLE_IPV4
 v4_sbin_links  = iptables-legacy iptables-legacy-restore iptables-legacy-save \
 		 iptables iptables-restore iptables-save
Index: iptables-1.8.10/iptables/xtables-legacy-multi.c
===================================================================
--- iptables-1.8.10.orig/iptables/xtables-legacy-multi.c
+++ iptables-1.8.10/iptables/xtables-legacy-multi.c
@@ -28,8 +28,10 @@ static const struct subcommand multi_sub
 
 
 #endif
+#ifdef ENABLE_XML
 	{"iptables-xml",        iptables_xml_main},
 	{"xml",                 iptables_xml_main},
+#endif
 #ifdef ENABLE_IPV6
 	{"ip6tables",           ip6tables_main},
 	{"main6",               ip6tables_main},
Index: iptables-1.8.10/iptables/xtables-nft-multi.c
===================================================================
--- iptables-1.8.10.orig/iptables/xtables-nft-multi.c
+++ iptables-1.8.10/xiptables/tables-nft-multi.c
@@ -6,8 +6,10 @@
 #include "xtables-multi.h"
 
 static const struct subcommand multi_subcommands[] = {
+#ifdef ENABLE_XML
 	{"iptables-xml",		iptables_xml_main},
 	{"xml",				iptables_xml_main},
+#endif
 	{"iptables",			xtables_ip4_main},
 	{"iptables-nft",		xtables_ip4_main},
 	{"main4",			xtables_ip4_main},
