--- sqlite-autoconf-3071602/sqlite3.c.orig	2013-05-06 15:52:48.980797533 +1000
+++ sqlite-autoconf-3071602/sqlite3.c	2013-05-06 15:54:07.470420407 +1000
@@ -22772,8 +22772,10 @@
 */
 #if !defined(HAVE_POSIX_FALLOCATE) \
       && (_XOPEN_SOURCE >= 600 || _POSIX_C_SOURCE >= 200112L)
+#if !defined(__UCLIBC__)
 # define HAVE_POSIX_FALLOCATE 1
 #endif
+#endif
 
 /*
 ** There are various methods for file locking used for concurrency
