diff -Naur libmnl-1.0.3.org/src/internal.h libmnl-1.0.3/src/internal.h
--- libmnl-1.0.3.org/src/internal.h	2011-02-03 20:40:52.000000000 +1000
+++ libmnl-1.0.3/src/internal.h	2014-02-21 13:08:30.000000000 +1000
@@ -6,7 +6,7 @@
 #	define __visible	__attribute__((visibility("default")))
 #	define EXPORT_SYMBOL(x)	typeof(x) (x) __visible
 #else
-#	define EXPORT_SYMBOL
+#	define EXPORT_SYMBOL(x)
 #endif
 
 #endif
