diff -Naur busybox-1.31.0.org/coreutils/sync.c busybox-1.31.0/coreutils/sync.c
--- busybox-1.31.0.org/coreutils/sync.c	2019-06-10 19:09:31.000000000 +1000
+++ busybox-1.31.0/coreutils/sync.c	2019-07-02 13:50:05.417882112 +1000
@@ -39,6 +39,7 @@
 //usage:     "\n	-f	Sync filesystems underlying FILEs"
 //usage:    )
 
+#include <config/autoconf.h>
 #include "libbb.h"
 
 /* This is a NOFORK applet. Be very careful! */
@@ -50,6 +51,7 @@
 	enum {
 		OPT_DATASYNC = (1 << 0),
 		OPT_SYNCFS   = (1 << 1),
+		OPT_FLATFSD  = (1 << 2),
 	};
 
 	ret = EXIT_SUCCESS;
@@ -62,6 +64,14 @@
 			ret = EXIT_FAILURE;
 			goto next;
 		}
+# if defined(CONFIG_USER_FLATFSD_FLATFSD)
+		if (opts & OPT_FLATFSD) {
+			/*
+			 * Call out to flatfsd to flush the flash filesystem.
+			 */
+			system("exec flatfsd -s");
+		} else
+# endif
 # if ENABLE_FEATURE_SYNC_FANCY
 		if (opts & OPT_SYNCFS) {
 			/*
@@ -89,12 +99,23 @@
 int sync_main(int argc UNUSED_PARAM, char **argv IF_NOT_DESKTOP(UNUSED_PARAM))
 {
 # if !ENABLE_FEATURE_SYNC_FANCY
+	unsigned opts = 0;
+#ifdef CONFIG_USER_FLATFSD_FLATFSD
+	opts = getopt32(argv, "F");
+
+	if (opts) {
+		system("exec flatfsd -s");
+		opts = 0;
+	}
+#endif
+
 	/* coreutils-6.9 compat */
-	bb_warn_ignoring_args(argv[1]);
+	if (opts)
+		bb_warn_ignoring_args(argv[1]);
 	sync();
 	return EXIT_SUCCESS;
 # else
-	unsigned opts = getopt32(argv, "^" "df" "\0" "d--f:f--d");
+	unsigned opts = getopt32(argv, "^" "dfF" "\0" "d--f:f--d");
 	argv += optind;
 	if (!argv[0]) {
 		sync();
diff -Naur busybox-1.31.0.org/init/halt.c busybox-1.31.0/init/halt.c
--- busybox-1.31.0.org/init/halt.c	2018-12-31 01:14:20.000000000 +1000
+++ busybox-1.31.0/init/halt.c	2019-07-02 13:12:39.519850841 +1000
@@ -93,6 +93,7 @@
 
 #include "libbb.h"
 #include "reboot.h"
+#include <config/autoconf.h>
 
 #if ENABLE_FEATURE_WTMP
 #include <sys/utsname.h>
@@ -169,6 +170,13 @@
 		RB_POWER_OFF,
 		RB_AUTOBOOT
 	};
+#ifdef CONFIG_USER_FLATFSD_FLATFSD
+	static const char *flatfsd[] = {
+		"exec /bin/flatfsd -H",
+		"exec /bin/flatfsd -P",
+		"exec /bin/flatfsd -b",
+	};
+#endif
 	static const smallint signals[] = { SIGUSR1, SIGUSR2, SIGTERM };
 
 	int delay = 0;
@@ -196,6 +204,15 @@
 
 	sleep(delay);
 
+#ifdef CONFIG_USER_FLATFSD_FLATFSD
+	if (!(flags & 4)) {
+		/* Ask flatfsd to halt us safely */
+		if (system(flatfsd[which]) != -1) {
+			exit(0);
+		}
+	}
+#endif
+
 	write_wtmp();
 
 	if (flags & 8) /* -w */
diff -Naur busybox-1.31.0.org/sysklogd/syslogd.c busybox-1.31.0/sysklogd/syslogd.c
--- busybox-1.31.0.org/sysklogd/syslogd.c	2019-06-10 19:09:31.000000000 +1000
+++ busybox-1.31.0/sysklogd/syslogd.c	2019-07-02 13:12:39.519850841 +1000
@@ -384,6 +384,13 @@
 }
 
 #if ENABLE_FEATURE_SYSLOGD_CFG
+
+#ifdef CONFIG_USER_FLATFSD_ETC_CONFIG
+#define DEFAULT_CONFIG_FILE "/etc/config/syslog.conf"
+#else
+#define DEFAULT_CONFIG_FILE "/etc/syslog.conf"
+#endif
+
 static void parse_syslogdcfg(const char *file)
 {
 	char *t;
@@ -394,7 +401,7 @@
 	char *tok[3];
 	parser_t *parser;
 
-	parser = config_open2(file ? file : "/etc/syslog.conf",
+	parser = config_open2(file ? file : DEFAULT_CONFIG_FILE,
 				file ? xfopen_for_read : fopen_for_read);
 	if (!parser)
 		/* didn't find default /etc/syslog.conf */
@@ -536,7 +543,7 @@
 
  cfgerr:
 	bb_error_msg_and_die("error in '%s' at line %d",
-			file ? file : "/etc/syslog.conf",
+			file ? file : DEFAULT_CONFIG_FILE,
 			parser->lineno);
 }
 #endif
