diff -Naur busybox-1.34.1.org/init/init.c busybox-1.34.1/init/init.c
--- busybox-1.34.1.org/init/init.c	2018-12-31 01:14:20.000000000 +1000
+++ busybox-1.34.1/init/init.c	2019-07-02 14:14:24.967531691 +1000
@@ -118,6 +118,13 @@
 //config:	be they parsed or ignored by init.
 //config:	The original command-line used to launch init can then be
 //config:	retrieved in /proc/1/cmdline on Linux, for example.
+//config:
+//config:config FEATURE_ETC_RC
+//config:	bool "Run /etc/rc before loading /etc/inittab"
+//config:	default n
+//config:	depends on INIT
+//config:	help
+//config:	Legacy in case inittab lives in an unmounted filesystem.
 
 //applet:IF_INIT(APPLET(init, BB_DIR_SBIN, BB_SUID_DROP))
 //applet:IF_LINUXRC(APPLET_ODDNAME(linuxrc, init, BB_DIR_ROOT, BB_SUID_DROP, linuxrc))
@@ -1118,6 +1125,19 @@
 	message(L_CONSOLE | L_LOG, "init started: %s", bb_banner);
 #endif
 
+#if ENABLE_FEATURE_ETC_RC
+	{
+		pid_t pid;
+		pid = vfork();
+		if (pid == 0) { /* child */
+			close(0);
+			execl("/bin/sh", "sh", "/etc/rc", NULL);
+			_exit(0);
+		}
+		waitfor(pid);
+	}
+#endif
+
 	/* Check if we are supposed to be in single user mode */
 	if (argv[1]
 	 && (strcmp(argv[1], "single") == 0 || strcmp(argv[1], "-s") == 0 || LONE_CHAR(argv[1], '1'))
