diff -Naur busybox-1.36.1.org/miscutils/watchdog.c busybox-1.36.1/miscutils/watchdog.c
--- busybox-1.36.1.org/miscutils/watchdog.c	2019-06-10 19:09:31.000000000 +1000
+++ busybox-1.36.1/miscutils/watchdog.c	2019-07-02 14:40:30.725595960 +1000
@@ -97,6 +97,14 @@
 #endif
 }
 
+static void watchdog_safe_terminate(int unused UNUSED_PARAM) NORETURN;
+static void watchdog_safe_terminate(int unused UNUSED_PARAM)
+{
+	write(3, "T", 1); /* Magic safe terminate (ie, a longer timeout) */
+	close(3);
+	exit(0);
+}
+
 int watchdog_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
 int watchdog_main(int argc UNUSED_PARAM, char **argv)
 {
@@ -112,6 +120,7 @@
 	unsigned htimer_duration = 60000; /* reboots after N ms if not restarted */
 	char *st_arg;
 	char *ht_arg;
+	int fd;
 
 #define OPT_FOREGROUND  (1 << 0)
 #define OPT_STIMER      (1 << 1)
@@ -137,7 +146,22 @@
 	if (opts & OPT_STIMER)
 		stimer_duration = xatou_sfx(st_arg, suffixes);
 
-	bb_signals(BB_FATAL_SIGS, shutdown_on_signal);
+	fd = open("/proc/self/oom_score_adj", O_WRONLY);
+	if (fd >= 0) {
+		write(fd, "-1000", 5);
+		close(fd);
+	} else {
+		fd = open("/proc/self/oom_adj", O_WRONLY);
+		if (fd >= 0) {
+			write(fd, "-17", 3);
+			close(fd);
+		}
+	}
+
+	signal(SIGHUP, shutdown_on_signal);
+	signal(SIGINT, shutdown_on_signal);
+	signal(SIGTERM, watchdog_safe_terminate);
+	//bb_signals(BB_FATAL_SIGS, shutdown_on_signal);
 
 	watchdog_open(argv[optind]);
 
