--- musl-1.2.3/src/time/__tz.c.orig	2022-04-08 03:12:40.000000000 +1000
+++ musl-1.2.3/src/time/__tz.c	14:05:45.296238460 +1000
@@ -4,6 +4,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <sys/mman.h>
+#include <sys/stat.h>
 #include <ctype.h>
 #include "libc.h"
 #include "lock.h"
@@ -131,12 +132,25 @@
 	size_t i;
 	static const char search[] =
 		"/usr/share/zoneinfo/\0/share/zoneinfo/\0/etc/zoneinfo/\0";
+	static time_t prev_mtime;
+	struct stat st;
 
 	s = getenv("TZ");
-	if (!s) s = "/etc/localtime";
+	if (!s) s = TZ_FILE_PATH;
+
+	/* 
+	 * If s is not a file (or doesn't exist yet), the stat command will fail.
+	 * In this case, make sure the mtime value is 0.
+	 */
+	if (stat(s, &st) == -1)
+		st.st_mtime = 0;
+
 	if (!*s) s = __utc;
 
-	if (old_tz && !strcmp(s, old_tz)) return;
+	if (st.st_mtime == prev_mtime && old_tz && !strcmp(s, old_tz))
+		return;
+
+	prev_mtime = st.st_mtime;
 
 	for (i=0; i<5; i++) r0[i] = r1[i] = 0;
 
@@ -172,7 +186,7 @@
 	if (!posix_form) {
 		if (*s == ':') s++;
 		if (*s == '/' || *s == '.') {
-			if (!libc.secure || !strcmp(s, "/etc/localtime"))
+			if (!libc.secure || !strcmp(s, TZ_FILE_PATH))
 				map = __map_file(s, &map_size);
 		} else {
 			size_t l = strlen(s);
@@ -188,6 +202,23 @@
 		}
 		if (!map) s = __utc;
 	}
+
+	if (map) {
+		if (map_size > 0 && map_size < sizeof(buf)) {
+			char *c;
+			memcpy(buf, map, map_size);
+			buf[map_size] = 0;
+			if ((c = strchr(buf, '\n')))
+				*c = 0;
+			s = buf;
+		} else {
+			s = __utc;
+		}
+		__munmap((void *)map, map_size);
+		map = 0;
+	}
+
+#if 0
 	if (map && (map_size < 44 || memcmp(map, "TZif", 4))) {
 		__munmap((void *)map, map_size);
 		map = 0;
@@ -235,6 +266,7 @@
 			return;
 		}
 	}
+#endif
 
 	if (!s) s = __utc;
 	getname(std_name, &s);
