diff -Naur a/tools/omfile.h b/tools/omfile.h
--- a/tools/omfile.h	2017-12-21 22:09:39.000000000 +1000
+++ b/tools/omfile.h	2018-02-21 08:10:57.467177356 +1000
@@ -32,7 +32,7 @@
  * other way to have the functionality (well, one way would be to go through the
  * globals, but that seems not yet justified. -- rgerhards, 2010-03-01
  */
-uchar	*pszFileDfltTplName;
+extern uchar	*pszFileDfltTplName;
 #endif /* #ifndef OMFILE_H_INCLUDED */
 /* vi:set ai:
  */
