--- parted-3.3/libparted/labels/gpt.c.save	2018-12-20 17:11:02.039393265 +1000
+++ parted-3.3/libparted/labels/gpt.c	2018-12-20 17:11:04.471393260 +1000
@@ -761,6 +761,7 @@
 
   if (last_usable < last_usable_if_grown)
     {
+#ifndef ALWAYS_FIX_GPT
       PedExceptionOption q;
 
       q = ped_exception_throw
@@ -774,6 +775,7 @@
          (uint64_t) (last_usable_if_grown - last_usable));
 
       if (q == PED_EXCEPTION_FIX)
+#endif
         {
           last_usable = last_usable_if_grown;
           /* clear the old backup gpt header */
