From 156aea9170e82768c582bcb47ea0331b8c5758a5 Mon Sep 17 00:00:00 2001
From: David McCullough <David.McCullough@digi.com>
Date: Tue, 25 Jan 2022 17:35:39 +0100
Subject: [PATCH 05/10] Adding 'virtualuser' config parameter

---
 auth.c     | 12 +++++++++---
 loginrec.c |  8 +++++++-
 servconf.c | 15 +++++++++++++++
 servconf.h |  2 ++
 4 files changed, 33 insertions(+), 4 deletions(-)

diff --git a/auth.c b/auth.c
index 3b380d9bb..9e9c9cee2 100644
--- a/auth.c
+++ b/auth.c
@@ -485,7 +485,7 @@ getpwnamallow(struct ssh *ssh, const char *user)
 	aix_setauthdb(user);
 #endif
 
-	pw = getpwnam(user);
+	pw = getpwnam(options.virtual_user ?: user);
 
 #if defined(_AIX) && defined(HAVE_SETAUTHDB)
 	aix_restoreauthdb();
@@ -519,8 +519,14 @@ getpwnamallow(struct ssh *ssh, const char *user)
 		auth_close(as);
 #endif
 #endif
-	if (pw != NULL)
-		return (pwcopy(pw));
+	if (pw != NULL) {
+		pw = pwcopy(pw);
+		if (pw && options.virtual_user && strcmp(options.virtual_user, pw->pw_name) == 0) {
+			free(pw->pw_name);
+			pw->pw_name = xstrdup(user);
+		}
+		return (pw);
+	}
 	return (NULL);
 }
 
diff --git a/loginrec.c b/loginrec.c
index 4f2149958..cc0059854 100644
--- a/loginrec.c
+++ b/loginrec.c
@@ -169,6 +169,11 @@
 #include <time.h>
 #include <unistd.h>
 
+#include "openbsd-compat/sys-queue.h"
+#include "channels.h"
+#include "misc.h"
+#include "log.h"
+#include "servconf.h"
 #include "xmalloc.h"
 #include "sshkey.h"
 #include "hostfile.h"
@@ -215,6 +220,7 @@ int wtmp_get_entry(struct logininfo *li);
 int wtmpx_get_entry(struct logininfo *li);
 
 extern struct sshbuf *loginmsg;
+extern ServerOptions options;
 
 /* pick the shortest string */
 #define MIN_SIZEOF(s1,s2) (sizeof(s1) < sizeof(s2) ? sizeof(s1) : sizeof(s2))
@@ -385,7 +391,7 @@ login_init_entry(struct logininfo *li, pid_t pid, const char *username,
 
 	if (username) {
 		strlcpy(li->username, username, sizeof(li->username));
-		pw = getpwnam(li->username);
+		pw = getpwnam(options.virtual_user ?: li->username);
 		if (pw == NULL) {
 			fatal("%s: Cannot find user \"%s\"", __func__,
 			    li->username);
diff --git a/servconf.c b/servconf.c
index 4b434909a..0013d8791 100644
--- a/servconf.c
+++ b/servconf.c
@@ -190,6 +190,7 @@ initialize_server_options(ServerOptions *options)
 	options->ip_qos_interactive = -1;
 	options->ip_qos_bulk = -1;
 	options->version_addendum = NULL;
+	options->virtual_user = NULL;
 	options->fingerprint_hash = -1;
 	options->disable_forwarding = -1;
 	options->expose_userauth_info = -1;
@@ -529,6 +530,7 @@ typedef enum {
 	sAuthenticationMethods, sHostKeyAgent, sPermitUserRC,
 	sStreamLocalBindMask, sStreamLocalBindUnlink,
 	sAllowStreamLocalForwarding, sFingerprintHash, sDisableForwarding,
+	sVirtualUser,
 	sExposeAuthInfo, sRDomain, sPubkeyAuthOptions, sSecurityKeyProvider,
 	sRequiredRSASize, sChannelTimeout, sUnusedConnectionTimeout,
 	sDeprecated, sIgnore, sUnsupported
@@ -686,6 +688,7 @@ static struct {
 	{ "allowstreamlocalforwarding", sAllowStreamLocalForwarding, SSHCFG_ALL },
 	{ "fingerprinthash", sFingerprintHash, SSHCFG_GLOBAL },
 	{ "disableforwarding", sDisableForwarding, SSHCFG_ALL },
+	{ "virtualuser", sVirtualUser, SSHCFG_ALL },
 	{ "exposeauthinfo", sExposeAuthInfo, SSHCFG_ALL },
 	{ "rdomain", sRDomain, SSHCFG_ALL },
 	{ "casignaturealgorithms", sCASignatureAlgorithms, SSHCFG_ALL },
@@ -2525,6 +2528,17 @@ process_server_config_line_depth(ServerOptions *options, char *line,
 			options->fingerprint_hash = value;
 		break;
 
+	case sVirtualUser:
+		charptr = &options->virtual_user;
+
+		arg = argv_next(&ac, &av);
+		if (!arg || *arg == '\0')
+			fatal("%s line %d: missing user name.",
+			    filename, linenum);
+		if (*activep && *charptr == NULL)
+			*charptr = xstrdup(arg);
+		break;
+
 	case sExposeAuthInfo:
 		intptr = &options->expose_userauth_info;
 		goto parse_flag;
@@ -3164,6 +3178,7 @@ dump_config(ServerOptions *o)
 	dump_cfg_string(sHostbasedAcceptedAlgorithms, o->hostbased_accepted_algos);
 	dump_cfg_string(sHostKeyAlgorithms, o->hostkeyalgorithms);
 	dump_cfg_string(sPubkeyAcceptedAlgorithms, o->pubkey_accepted_algos);
+	dump_cfg_string(sVirtualUser, o->virtual_user);
 #if defined(__OpenBSD__) || defined(HAVE_SYS_SET_PROCESS_RDOMAIN)
 	dump_cfg_string(sRDomain, o->routing_domain);
 #endif
diff --git a/servconf.h b/servconf.h
index ed7b72e8e..c7b1e1266 100644
--- a/servconf.h
+++ b/servconf.h
@@ -223,6 +223,8 @@ typedef struct {
 	u_int	num_auth_methods;
 	char   **auth_methods;
 
+	char   *virtual_user;
+
 	int	fingerprint_hash;
 	int	expose_userauth_info;
 	u_int64_t timing_secret;
2.30.2

