Index: openssh-6.6p1/auth.c
===================================================================
--- openssh-6.6p1.orig/auth.c
+++ openssh-6.6p1/auth.c
@@ -603,7 +603,7 @@
 	aix_setauthdb(user);
 #endif
 
-	pw = getpwnam(user);
+	pw = getpwnam(options.virtual_user ?: user);
 
 #if defined(_AIX) && defined(HAVE_SETAUTHDB)
 	aix_restoreauthdb();
@@ -650,8 +650,14 @@
 		auth_close(as);
 #endif
 #endif
-	if (pw != NULL)
-		return (pwcopy(pw));
+	if (pw != NULL) {
+		pw = pwcopy(pw);
+		if (pw && options.virtual_user && strcmp(options.virtual_user, pw->pw_name) == 0) {
+			free(pw->pw_name);
+			pw->pw_name = xstrdup(user);
+		}
+		return (pw);
+	}
 	return (NULL);
 }
 
--- openssh-6.8p1/servconf.c.orig	2015-03-17 15:49:20.000000000 +1000
+++ openssh-6.8p1/servconf.c	2015-03-19 10:22:00.704516821 +1000
@@ -162,6 +162,7 @@
 	options->ip_qos_interactive = -1;
 	options->ip_qos_bulk = -1;
 	options->version_addendum = NULL;
+	options->virtual_user = NULL;
 	options->fingerprint_hash = -1;
 	options->disable_forwarding = -1;
 }
@@ -492,6 +493,7 @@
 	sAuthenticationMethods, sHostKeyAgent, sPermitUserRC,
 	sStreamLocalBindMask, sStreamLocalBindUnlink,
 	sAllowStreamLocalForwarding, sFingerprintHash, sDisableForwarding,
+	sVirtualUser,
 	sExposeAuthInfo, sRDomain,
 	sDeprecated, sIgnore, sUnsupported
 } ServerOpCodes;
@@ -638,6 +640,7 @@
 	{ "allowstreamlocalforwarding", sAllowStreamLocalForwarding, SSHCFG_ALL },
 	{ "fingerprinthash", sFingerprintHash, SSHCFG_GLOBAL },
 	{ "disableforwarding", sDisableForwarding, SSHCFG_ALL },
+	{ "virtualuser", sVirtualUser, SSHCFG_ALL },
 	{ "exposeauthinfo", sExposeAuthInfo, SSHCFG_ALL },
 	{ "rdomain", sRDomain, SSHCFG_ALL },
 	{ NULL, sBadOption, 0 }
@@ -1743,6 +1746,17 @@
 			options->fingerprint_hash = value;
 		break;
 
+	case sVirtualUser:
+		charptr = &options->virtual_user;
+
+		arg = strdelim(&cp);
+		if (!arg || *arg == '\0')
+			fatal("%s line %d: missing user name.",
+			    filename, linenum);
+		if (*activep && *charptr == NULL)
+			*charptr = xstrdup(arg);
+		break;
+
 	case sDeprecated:
 		logit("%s line %d: Deprecated option %s",
 		    filename, linenum, arg);
@@ -2173,6 +2187,7 @@
 	    o->hostbased_key_types : KEX_DEFAULT_PK_ALG);
 	dump_cfg_string(sPubkeyAcceptedKeyTypes, o->pubkey_key_types ?
 	    o->pubkey_key_types : KEX_DEFAULT_PK_ALG);
+	dump_cfg_string(sVirtualUser, o->virtual_user);
 	dump_cfg_string(sRDomain, o->routing_domain);
 
 	/* string arguments requiring a lookup */
--- openssh-6.8p1/servconf.h.orig	2015-03-19 10:15:51.640525369 +1000
+++ openssh-6.8p1/servconf.h	2015-03-19 10:23:09.460515228 +1000
@@ -206,6 +206,8 @@
 	u_int	num_auth_methods;
 	char   **auth_methods;
 
+	char   *virtual_user;
+
 	int	fingerprint_hash;
 	int	expose_userauth_info;
 	u_int64_t timing_secret;
Index: openssh-6.6p1/loginrec.c
===================================================================
--- openssh-6.6p1.orig/loginrec.c
+++ openssh-6.6p1/loginrec.c
@@ -167,6 +167,11 @@
 #include <time.h>
 #include <unistd.h>
 
+#include "openbsd-compat/sys-queue.h"
+#include "channels.h"
+#include "misc.h"
+#include "log.h"
+#include "servconf.h"
 #include "xmalloc.h"
 #include "sshkey.h"
 #include "hostfile.h"
@@ -212,6 +213,7 @@
 int wtmpx_get_entry(struct logininfo *li);
 
 extern struct sshbuf *loginmsg;
+extern ServerOptions options;
 
 /* pick the shortest string */
 #define MIN_SIZEOF(s1,s2) (sizeof(s1) < sizeof(s2) ? sizeof(s1) : sizeof(s2))
@@ -378,7 +380,7 @@
 
 	if (username) {
 		strlcpy(li->username, username, sizeof(li->username));
-		pw = getpwnam(li->username);
+		pw = getpwnam(options.virtual_user ?: li->username);
 		if (pw == NULL) {
 			fatal("%s: Cannot find user \"%s\"", __func__,
 			    li->username);
