diff -Naur libnetfilter_cttimeout-1.0.0.org/src/internal.h libnetfilter_cttimeout-1.0.0/src/internal.h
--- libnetfilter_cttimeout-1.0.0.org/src/internal.h	2012-02-29 12:19:01.000000000 +1000
+++ libnetfilter_cttimeout-1.0.0/src/internal.h	2014-02-21 13:19:39.000000000 +1000
@@ -6,7 +6,7 @@
 #	define __visible	__attribute__((visibility("default")))
 #	define EXPORT_SYMBOL(x)	typeof(x) (x) __visible
 #else
-#	define EXPORT_SYMBOL
+#	define EXPORT_SYMBOL(x)
 #endif
 
 #endif
