From: David Escalona <david.escalona@digi.com>
Date: Fri, 4 May 2018 18:54:48 +0200
Subject: [PATCH 4/9] reallocarray: add reallocarray function definition

Signed-off-by: David Escalona <david.escalona@digi.com>
---
diff --git a/libraries/libapparmor/src/private.c b/libraries/libapparmor/src/private.c
index bfafdb8..cbc3d84 100644
--- a/libraries/libapparmor/src/private.c
+++ b/libraries/libapparmor/src/private.c
@@ -75,6 +75,30 @@ static struct ignored_suffix_t ignored_suffixes[] = {
 	{ NULL, 0, 0 }
 };
 
+#ifdef NEED_REALLOCARRAY
+
+#ifdef HAVE_STDINT_H
+#include <stdint.h>
+#endif
+
+/*
+ * This is sqrt(SIZE_MAX+1), as s1*s2 <= SIZE_MAX
+ * if both s1 < MUL_NO_OVERFLOW and s2 < MUL_NO_OVERFLOW
+ */
+#define MUL_NO_OVERFLOW	((size_t)1 << (sizeof(size_t) * 4))
+
+void *
+reallocarray(void *optr, size_t nmemb, size_t size)
+{
+	if ((nmemb >= MUL_NO_OVERFLOW || size >= MUL_NO_OVERFLOW) &&
+	    nmemb > 0 && SIZE_MAX / nmemb < size) {
+		errno = ENOMEM;
+		return NULL;
+	}
+	return realloc(optr, size * nmemb);
+}
+#endif
+
 #define DEBUG_ENV_VAR	"LIBAPPARMOR_DEBUG"
 
 void print_error(bool honor_env_var, const char *ident, const char *fmt, ...)
