--- apparmor-2.13/parser/parser_alias.c.org	2019-05-31 22:58:35.315083432 +1000
+++ apparmor-2.13/parser/parser_alias.c	2019-05-31 22:57:45.505174725 +1000
@@ -25,6 +25,8 @@
 #include "parser.h"
 #include "profile.h"
 
+typedef int (*comparison_fn_t)(const void *, const void *);
+
 struct alias_rule {
 	char *from;
 	char *to;
--- apparmor-2.13/parser/parser_symtab.c.org	2019-05-31 22:58:40.486866462 +1000
+++ apparmor-2.13/parser/parser_symtab.c	2019-05-31 22:56:55.851262651 +1000
@@ -25,6 +25,9 @@
 #include "immunix.h"
 #include "parser.h"
 
+typedef int (*comparison_fn_t)(const void *, const void *);
+typedef void (*__free_fn_t)(void *);
+
 enum var_type {
 	sd_boolean,
 	sd_set,
