Index: ubusd_event.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/ubusd_event.c b/ubusd_event.c
--- a/ubusd_event.c	(revision 2f793a4eb063128d80fc46ada243d2aad47fd1fe)
+++ b/ubusd_event.c	(revision 2bebf93cd3343fe49f22a05ef935e460d2d44f67)
@@ -63,6 +63,9 @@
 	bool partial = false;
 	int len;
 
+	if (!msg)
+		return UBUS_STATUS_INVALID_ARGUMENT;
+
 	blobmsg_parse(evr_policy, EVREG_LAST, attr, blob_data(msg), blob_len(msg));
 	if (!attr[EVREG_OBJECT] || !attr[EVREG_PATTERN])
 		return UBUS_STATUS_INVALID_ARGUMENT;
@@ -209,6 +212,9 @@
 	struct blob_attr *attr[EVMSG_LAST];
 	const char *id;
 
+	if (!msg)
+		return UBUS_STATUS_INVALID_ARGUMENT;
+
 	blobmsg_parse(ev_policy, EVMSG_LAST, attr, blob_data(msg), blob_len(msg));
 	if (!attr[EVMSG_ID] || !attr[EVMSG_DATA])
 		return UBUS_STATUS_INVALID_ARGUMENT;
