diff --git a/cli.c b/cli.c
index 6bfb44e..c23aa50 100644
--- a/cli.c
+++ b/cli.c
@@ -57,6 +57,7 @@ static void receive_list_result(struct ubus_context *ctx, struct ubus_object_dat
 	}
 
 	printf("'%s' @%08x\n", obj->path, obj->id);
+	fflush(stdout);
 
 	if (!obj->signature)
 		return;
@@ -66,6 +67,7 @@ static void receive_list_result(struct ubus_context *ctx, struct ubus_object_dat
 		printf("\t%s\n", s);
 		free(s);
 	}
+	fflush(stdout);
 }
 
 static void receive_call_result_data(struct ubus_request *req, int type, struct blob_attr *msg)
@@ -76,6 +78,7 @@ static void receive_call_result_data(struct ubus_request *req, int type, struct
 
 	str = blobmsg_format_json_indent(msg, true, simple_output ? -1 : 0);
 	printf("%s\n", str);
+	fflush(stdout);
 	free(str);
 }
 
@@ -86,6 +89,7 @@ static void receive_event(struct ubus_context *ctx, struct ubus_event_handler *e
 
 	str = blobmsg_format_json(msg, true);
 	printf("{ \"%s\": %s }\n", type, str);
+	fflush(stdout);
 	free(str);
 }
 
