--- strongswan-5.7.1/src/libstrongswan/settings/settings_lexer.l.orig	2018-11-27 12:02:43.437689778 +1000
+++ strongswan-5.7.1/src/libstrongswan/settings/settings_lexer.l	2018-11-27 12:01:58.630368873 +1000
@@ -52,7 +52,7 @@
 /* state used to scan references */
 %x ref
 /* state used to scan values */
-%x val
+%x scanval
 /* state used to scan include file patterns */
 %x inc
 /* state used to scan quoted strings */
@@ -79,7 +79,7 @@
 }
 
 "="						{
-	yy_push_state(val, yyscanner);
+	yy_push_state(scanval, yyscanner);
 	return yytext[0];
 }
 
@@ -116,7 +116,7 @@
 	}
 }
 
-<val>{
+<scanval>{
 	\r					/* just ignore these */
 	[\t ]+
 	<<EOF>>				|
