--- nss-3.17.4/nspr/configure.in.LSW	2015-09-04 15:36:25.721646473 -0400
+++ nss-3.17.4/nspr/configure.in	2015-09-04 15:40:53.330624068 -0400
@@ -31,7 +31,7 @@
 USE_IPV6=
 USE_MDUPDATE=
 _MACOSX_DEPLOYMENT_TARGET=
-_OPTIMIZE_FLAGS=-O
+_OPTIMIZE_FLAGS=
 _DEBUG_FLAGS=-g
 MOZ_DEBUG=1
 MOZ_OPTIMIZE=
@@ -331,18 +331,6 @@
 	        AC_DEFINE(MOZILLA_CLIENT)
 	    fi])
 
-AC_ARG_ENABLE(optimize,
-    [  --enable-optimize[=OPT] Enable code optimizations (ie. -O2) ],
-    [ if test "$enableval" != "no"; then
-          MOZ_OPTIMIZE=1
-          if test -n "$enableval" -a "$enableval" != "yes"; then
-            _OPTIMIZE_FLAGS=`echo $enableval | sed -e 's|\\\ | |g'`
-            _SAVE_OPTIMIZE_FLAGS=$_OPTIMIZE_FLAGS
-          fi
-      else
-          MOZ_OPTIMIZE=
-      fi ])
-
 AC_ARG_ENABLE(debug,
     [  --enable-debug[=DBG]    Enable debugging (using compiler flags DBG)],
     [ if test "$enableval" != "no"; then
@@ -1907,10 +1895,6 @@
             PR_MD_ASFILES=os_Linux_ppc.s
         fi
         ;;
-    m68k)
-        echo CFLAGS="$CFLAGS -m68020-60"
-        echo CXXFLAGS="$CXXFLAGS -m68020-60"
-        ;;
     esac    
     ;;
 
