diff -ru pam_ldap-185.orig/Makefile.am pam_ldap-185/Makefile.am
--- pam_ldap-185.orig/Makefile.am	2017-11-14 10:21:30.213859225 +1000
+++ pam_ldap-185/Makefile.am	2017-11-14 10:22:00.734070357 +1000
@@ -44,4 +44,4 @@
 	@$(NORMAL_UNINSTALL)
 
 vers.c: $(top_srcdir)/CVSVersionInfo.txt
-	CVSVERSIONDIR=$(top_srcdir) vers_string -v
+	CVSVERSIONDIR=$(top_srcdir) PERL5LIB=$(top_srcdir) ./vers_string -v
diff -ru pam_ldap-185.orig/Makefile.in pam_ldap-185/Makefile.in
--- pam_ldap-185.orig/Makefile.in	2009-11-06 20:29:34.000000000 +1000
+++ pam_ldap-185/Makefile.in	2017-11-14 10:22:05.598104310 +1000
@@ -611,7 +611,7 @@
 	@$(NORMAL_UNINSTALL)
 
 vers.c: $(top_srcdir)/CVSVersionInfo.txt
-	CVSVERSIONDIR=$(top_srcdir) vers_string -v
+	CVSVERSIONDIR=$(top_srcdir) PERL5LIB=$(top_srcdir) ./vers_string -v
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
 .NOEXPORT:
