--- netifd/bridge.c.orig	2021-07-28 14:46:58.265590695 +1000
+++ netifd/bridge.c	2021-07-28 14:49:17.193588645 +1000
@@ -34,6 +34,7 @@
 	BRIDGE_ATTR_BRIDGE_EMPTY,
 	BRIDGE_ATTR_MULTICAST_QUERIER,
 	BRIDGE_ATTR_HASH_MAX,
+	BRIDGE_ATTR_VLAN_FILTERING,
 	__BRIDGE_ATTR_MAX
 };
 
@@ -49,6 +50,7 @@
 	[BRIDGE_ATTR_BRIDGE_EMPTY] = { "bridge_empty", BLOBMSG_TYPE_BOOL },
 	[BRIDGE_ATTR_MULTICAST_QUERIER] = { "multicast_querier", BLOBMSG_TYPE_BOOL },
 	[BRIDGE_ATTR_HASH_MAX] = { "hash_max", BLOBMSG_TYPE_INT32 },
+	[BRIDGE_ATTR_VLAN_FILTERING] = { "vlan_filtering", BLOBMSG_TYPE_BOOL },
 };
 
 static const struct uci_blob_param_info bridge_attr_info[__BRIDGE_ATTR_MAX] = {
@@ -554,6 +556,7 @@
 
 	/* defaults */
 	cfg->stp = false;
+	cfg->vlan_filtering = false;
 	cfg->forward_delay = 2;
 	cfg->igmp_snoop = true;
 	cfg->multicast_querier = true;
@@ -564,6 +567,9 @@
 	if ((cur = tb[BRIDGE_ATTR_STP]))
 		cfg->stp = blobmsg_get_bool(cur);
 
+	if ((cur = tb[BRIDGE_ATTR_VLAN_FILTERING]))
+		cfg->vlan_filtering = blobmsg_get_bool(cur);
+
 	if ((cur = tb[BRIDGE_ATTR_FORWARD_DELAY]))
 		cfg->forward_delay = blobmsg_get_u32(cur);
 
--- netifd/system.h.orig	2021-07-28 14:46:31.017591097 +1000
+++ netifd/system.h	2021-07-28 14:49:52.857588119 +1000
@@ -49,6 +49,7 @@
 struct bridge_config {
 	enum bridge_opt flags;
 	bool stp;
+	bool vlan_filtering;
 	bool igmp_snoop;
 	bool multicast_querier;
 	unsigned short priority;
--- netifd/system-linux.c.orig	2021-07-29 08:32:46.218970304 +1000
+++ netifd/system-linux.c	2021-07-29 08:34:05.210966828 +1000
@@ -811,6 +811,9 @@
 	system_set_dev_sysctl("/sys/devices/virtual/net/%s/bridge/multicast_querier",
 		bridge->ifname, cfg->multicast_querier ? "1" : "0");
 
+	system_set_dev_sysctl("/sys/devices/virtual/net/%s/bridge/vlan_filtering",
+		bridge->ifname, cfg->vlan_filtering ? "1" : "0");
+
 	snprintf(buf, sizeof(buf), "%i", cfg->hash_max);
 	system_set_dev_sysctl("/sys/devices/virtual/net/%s/bridge/hash_max",
 		bridge->ifname, buf);
