Index: netifd/wireless.h
===================================================================
--- netifd.orig/wireless.h
+++ netifd/wireless.h
@@ -17,8 +17,8 @@
 #include <libubox/utils.h>
 #include "interface.h"
 
-struct vlist_tree wireless_devices;
-struct avl_tree wireless_drivers;
+extern struct vlist_tree wireless_devices;
+extern struct avl_tree wireless_drivers;
 
 struct wireless_driver {
 	struct avl_node node;
