--- ../clone/netifd/scripts/netifd-wireless.sh.orig	2019-03-25 10:45:10.000000000 -0400
+++ netifd/scripts/netifd-wireless.sh	2021-03-29 19:09:03.778221108 -0400
@@ -198,7 +198,7 @@
 	wpa_pairwise=CCMP
 	case "$encryption" in
 		*tkip+aes|*tkip+ccmp|*aes+tkip|*ccmp+tkip) wpa_pairwise="CCMP TKIP";;
-		*aes|*ccmp) wpa_pairwise="CCMP";;
+		*aes|*ccmp|*sae) wpa_pairwise="CCMP";;
 		*tkip) wpa_pairwise="TKIP";;
 	esac
 
@@ -211,7 +211,7 @@
 	# wpa2/tkip+aes     => WPA2 RADIUS, CCMP+TKIP
 
 	case "$encryption" in
-		wpa2*|*psk2*)
+		wpa2*|*psk2*|wpa3*|*sae*|*owe*)
 			wpa=2
 		;;
 		*mixed*)
@@ -227,9 +227,25 @@
 	esac
 
 	case "$encryption" in
-		*psk*)
+		*psk2sae)
+			auth_type=psk2sae
+		;;
+		*owe)
+			auth_type=owe
+		;;
+		*psk)
+			auth_type=psk
+		;;
+		*psk2)
 			auth_type=psk
 		;;
+		*sae*)
+			auth_type=sae
+		;;
+		*wpa3*)
+			auth_type=eap192
+			wpa_pairwise="GCMP-256"
+		;;
 		*wpa*|*8021x*)
 			auth_type=eap
 		;;
