--- netifd.orig/interface.c	2017-10-17 14:46:23.623610331 +1000
+++ netifd/interface.c	2017-10-17 14:18:40.903583439 +1000
@@ -36,6 +36,7 @@
 	IFACE_ATTR_DNS,
 	IFACE_ATTR_DNS_SEARCH,
 	IFACE_ATTR_METRIC,
+	IFACE_ATTR_MTU,
 	IFACE_ATTR_INTERFACE,
 	IFACE_ATTR_IP6ASSIGN,
 	IFACE_ATTR_IP6HINT,
@@ -55,6 +56,7 @@
 	[IFACE_ATTR_DEFAULTROUTE] = { .name = "defaultroute", .type = BLOBMSG_TYPE_BOOL },
 	[IFACE_ATTR_PEERDNS] = { .name = "peerdns", .type = BLOBMSG_TYPE_BOOL },
 	[IFACE_ATTR_METRIC] = { .name = "metric", .type = BLOBMSG_TYPE_INT32 },
+	[IFACE_ATTR_MTU] = { .name = "mtu", .type = BLOBMSG_TYPE_INT32 },
 	[IFACE_ATTR_DNS] = { .name = "dns", .type = BLOBMSG_TYPE_ARRAY },
 	[IFACE_ATTR_DNS_SEARCH] = { .name = "dns_search", .type = BLOBMSG_TYPE_ARRAY },
 	[IFACE_ATTR_INTERFACE] = { .name = "interface", .type = BLOBMSG_TYPE_STRING },
@@ -779,6 +781,9 @@
 
 	if ((cur = tb[IFACE_ATTR_METRIC]))
 		iface->metric = blobmsg_get_u32(cur);
+
+	if ((cur = tb[IFACE_ATTR_MTU]))
+		iface->mtu = blobmsg_get_u32(cur);
 
 	if ((cur = tb[IFACE_ATTR_IP6ASSIGN]))
 		iface->assignment_length = blobmsg_get_u32(cur);
@@ -1076,6 +1082,7 @@
 void
 interface_update_complete(struct interface *iface)
 {
+	iface->proto_ip.iface->mtu = iface->mtu;
 	interface_ip_update_complete(&iface->proto_ip);
 }
 
@@ -1171,6 +1178,7 @@
 	interface_replace_dns(&if_old->config_ip, &if_new->config_ip);
 
 	UPDATE(metric, reload_ip);
+	UPDATE(mtu, reload_ip);
 	UPDATE(proto_ip.no_defaultroute, reload_ip);
 	UPDATE(ip4table, reload_ip);
 	UPDATE(ip6table, reload_ip);

--- netifd.orig/interface.h	2017-10-17 14:46:23.623610331 +1000
+++ netifd/interface.h	2017-10-17 14:18:40.903583439 +1000
@@ -141,6 +141,7 @@
 	struct vlist_tree host_routes;
 
 	int metric;
+	int mtu;
 	unsigned int ip4table;
 	unsigned int ip6table;
 

--- netifd.orig/interface-ip.c	2017-10-17 14:46:23.623610331 +1000
+++ netifd/interface-ip.c	2017-10-17 14:18:40.903583439 +1000
@@ -412,6 +412,9 @@
 
 	if (r1->metric != r2->metric)
 		return r1->metric - r2->metric;
+
+	if (r1->mtu != r2->mtu)
+		return r1->mtu - r2->mtu;
 
 	if (r1->flags != r2->flags)
 		return r2->flags - r1->flags;
@@ -462,6 +465,10 @@
 			route.sourcemask = 32;
 			memcpy(&route.source, &addr->addr, sizeof(route.addr));
 		}
+		if (iface->mtu != 0) {
+			route.mtu = iface->mtu;
+			route.flags |= DEVROUTE_MTU;
+		}
 		system_add_route(dev, &route);
 	} else {
 		system_del_route(dev, &route);
@@ -566,8 +573,8 @@
 				}
 			}
 
-			if (iface->metric)
+			if (iface->metric || (iface->mtu > 0))
 				interface_handle_subnet_route(iface, a_new, true);
 		}
 	}
 
--- netifd.orig/proto.c	2017-10-17 14:46:23.623610331 +1000
+++ netifd/proto.c	2017-10-17 14:18:40.903583439 +1000
@@ -271,6 +271,8 @@
 	route->mask = 0;
 	route->flags = (v6 ? DEVADDR_INET6 : DEVADDR_INET4);
 	route->metric = iface->metric;
+	route->mtu = iface->mtu;
+	route->flags |= DEVROUTE_MTU;
 
 	unsigned int table = (v6) ? iface->ip6table : iface->ip4table;
 	if (table) {

--- netifd.orig/proto-shell.c	2017-10-17 14:46:23.623610331 +1000
+++ netifd/proto-shell.c	2017-10-17 14:18:40.903583439 +1000
@@ -446,6 +446,7 @@
 	NOTIFY_LINK_UP,
 	NOTIFY_IFNAME,
 	NOTIFY_ADDR_EXT,
+	NOTIFY_MTU,
 	NOTIFY_ROUTES,
 	NOTIFY_ROUTES6,
 	NOTIFY_TUNNEL,
@@ -468,6 +469,7 @@
 	[NOTIFY_LINK_UP] = { .name = "link-up", .type = BLOBMSG_TYPE_BOOL },
 	[NOTIFY_IFNAME] = { .name = "ifname", .type = BLOBMSG_TYPE_STRING },
 	[NOTIFY_ADDR_EXT] = { .name = "address-external", .type = BLOBMSG_TYPE_BOOL },
+	[NOTIFY_MTU] = { .name = "mtu", .type = BLOBMSG_TYPE_INT32 },
 	[NOTIFY_ROUTES] = { .name = "routes", .type = BLOBMSG_TYPE_ARRAY },
 	[NOTIFY_ROUTES6] = { .name = "routes6", .type = BLOBMSG_TYPE_ARRAY },
 	[NOTIFY_TUNNEL] = { .name = "tunnel", .type = BLOBMSG_TYPE_TABLE },
@@ -505,6 +507,9 @@
 
 	if ((cur = tb[NOTIFY_KEEP]) != NULL)
 		keep = blobmsg_get_bool(cur);
+
+	if ((cur = tb[NOTIFY_MTU]) != NULL)
+		state->proto.iface->mtu = blobmsg_get_u32(cur);
 
 	if ((cur = tb[NOTIFY_ADDR_EXT]) != NULL) {
 		addr_ext = blobmsg_get_bool(cur);

diff -ur netifd.orig/scripts/netifd-proto.sh netifd/scripts/netifd-proto.sh
--- netifd.orig/scripts/netifd-proto.sh	2022-04-12 12:19:23.816366230 -0400
+++ netifd/scripts/netifd-proto.sh	2022-04-25 20:13:38.572385087 -0400
@@ -39,6 +39,7 @@
 
 	PROTO_KEEP=0
 	PROTO_INIT=1
+	PROTO_MTU=
 	PROTO_TUNNEL_OPEN=
 	PROTO_IPADDR=
 	PROTO_IP6ADDR=
@@ -104,13 +105,19 @@
 	append PROTO_DNS_SEARCH "$address"
 }
 
+proto_add_mtu() {
+	local mtu="$1"
+	append PROTO_MTU "$mtu"
+}
+
 proto_add_ipv4_address() {
 	local address="$1"
 	local mask="$2"
 	local broadcast="$3"
 	local ptp="$4"
+	local mtu="$5"
 
-	append PROTO_IPADDR "$address/$mask/$broadcast/$ptp"
+	append PROTO_IPADDR "$address/$mask/$broadcast/$ptp/$mtu"
 }
 
 proto_add_ipv6_address() {
@@ -130,8 +137,9 @@
 	local gw="$3"
 	local source="$4"
 	local metric="$5"
+	local mtu="$6"
 
-	append PROTO_ROUTE "$target/$mask/$gw/$metric///$source"
+	append PROTO_ROUTE "$target/$mask/$gw/$metric///$mtu/$source"
 }
 
 proto_add_ipv6_route() {
@@ -142,8 +150,9 @@
 	local valid="$5"
 	local source="$6"
 	local table="$7"
+	local mtu="$8"
 
-	append PROTO_ROUTE6 "$target/$mask/$gw/$metric/$valid/$table/$source"
+	append PROTO_ROUTE6 "$target/$mask/$gw/$metric/$valid/$table/$mtu/$source"
 }
 
 proto_add_ipv6_prefix() {
@@ -161,7 +170,7 @@
 
 _proto_push_ipv4_addr() {
 	local str="$1"
-	local address mask broadcast ptp
+	local address mask broadcast ptp mtu
 
 	address="${str%%/*}"
 	str="${str#*/}"
@@ -169,13 +178,16 @@
 	str="${str#*/}"
 	broadcast="${str%%/*}"
 	str="${str#*/}"
-	ptp="$str"
+	ptp="${str%%/*}"
+	str="${str#*/}"
+	mtu="$str"
 
 	json_add_object ""
 	json_add_string ipaddr "$address"
 	[ -n "$mask" ] && json_add_string mask "$mask"
 	[ -n "$broadcast" ] && json_add_string broadcast "$broadcast"
 	[ -n "$ptp" ] && json_add_string ptp "$ptp"
+	[ -n "$mtu" ] && json_add_int mtu "$mtu"
 	json_close_object
 }
 
@@ -223,7 +235,9 @@
 	str="${str#*/}"
 	local table="${str%%/*}"
 	str="${str#*/}"
-	local source="${str}"
+	local mtu="${str%%/*}"
+	str="${str#*/}"
+	local source="$str"
 
 	json_add_object ""
 	json_add_string target "$target"
@@ -233,6 +247,7 @@
 	[ -n "$valid" ] && json_add_int valid "$valid"
 	[ -n "$source" ] && json_add_string source "$source"
 	[ -n "$table" ] && json_add_string table "$table"
+	[ -n "$mtu" ] && json_add_int mtu "$mtu"
 	json_close_object
 }
 
@@ -261,6 +276,7 @@
 
 	proto_close_nested
 	json_add_boolean keep "$PROTO_KEEP"
+	json_add_int mtu "$PROTO_MTU"
 	_proto_push_array "ipaddr" "$PROTO_IPADDR" _proto_push_ipv4_addr
 	_proto_push_array "ip6addr" "$PROTO_IP6ADDR" _proto_push_ipv6_addr
 	_proto_push_array "routes" "$PROTO_ROUTE" _proto_push_route
  
--- netifd.orig/ubus.c	2017-10-17 14:46:23.623610331 +1000
+++ netifd/ubus.c	2017-10-17 14:18:40.903583439 +1000
@@ -113,11 +113,13 @@
 
 enum {
 	DI_NAME,
+	DI_MTU,
 	__DI_MAX
 };
 
 static const struct blobmsg_policy dynamic_policy[__DI_MAX] = {
 	[DI_NAME] = { .name = "name", .type = BLOBMSG_TYPE_STRING },
+	[DI_MTU] = { .name = "mtu", .type = BLOBMSG_TYPE_INT32 },
 };
 
 static int
@@ -696,6 +698,7 @@
 		if (iface->ip6table)
 			blobmsg_add_u32(&b, "ip6table", iface->ip6table);		  
 		blobmsg_add_u32(&b, "metric", iface->metric);
+		blobmsg_add_u32(&b, "mtu", iface->mtu);
 		blobmsg_add_u8(&b, "delegation", !iface->proto_ip.no_delegation);
 		a = blobmsg_open_array(&b, "ipv4-address");
 		interface_ip_dump_address_list(&iface->config_ip, false, true);

