Index: netifd/interface.c
===================================================================
--- netifd.orig/interface.c
+++ netifd/interface.c
@@ -243,6 +243,7 @@ mark_interface_down(struct interface *if
 	interface_ip_flush(&iface->proto_ip);
 	interface_flush_state(iface);
 	system_flush_routes();
+	netifd_ubus_interface_ip_event(iface);
 }
 
 void
@@ -677,6 +678,7 @@ interface_proto_cb(struct interface_prot
 		system_flush_routes();
 		iface->state = IFS_UP;
 		netifd_ubus_interface_state_event(iface);
+		netifd_ubus_interface_ip_event(iface);
 		iface->start_time = system_get_rtime();
 		interface_event(iface, IFEV_UP);
 		netifd_log_message(L_NOTICE, "Interface '%s' is now up\n", iface->name);
@@ -896,6 +898,7 @@ interface_set_l3_dev(struct interface *i
 		}
 		interface_ip_set_enabled(&iface->config_ip, enabled);
 	}
+	netifd_ubus_interface_ip_event(iface);
 }
 
 void
@@ -1188,6 +1191,7 @@ interface_change_config(struct interface
 		interface_ip_set_enabled(&if_old->proto_ip, false);
 		interface_ip_set_enabled(&if_old->proto_ip, proto_ip_enabled);
 		interface_ip_set_enabled(&if_old->config_ip, config_ip_enabled);
+		netifd_ubus_interface_ip_event(if_old);
 	}
 
 	interface_write_resolv_conf();
Index: netifd/ubus.c
===================================================================
--- netifd.orig/ubus.c
+++ netifd/ubus.c
@@ -1220,6 +1220,15 @@ netifd_ubus_interface_state_event(struct
 }
 
 void
+netifd_ubus_interface_ip_event(struct interface *iface)
+{
+	blob_buf_init(&b, 0);
+	blobmsg_add_string(&b, "action", "ip");
+	blobmsg_add_string(&b, "interface", iface->name);
+	ubus_send_event(ubus_ctx, "network.interface", b.head);
+}
+
+void
 netifd_ubus_interface_notify(struct interface *iface, bool up)
 {
 	const char *event = (up) ? "interface.update" : "interface.down";
Index: netifd/ubus.h
===================================================================
--- netifd.orig/ubus.h
+++ netifd/ubus.h
@@ -23,5 +23,6 @@ void netifd_ubus_remove_interface(struct
 void netifd_ubus_interface_event(struct interface *iface, bool up);
 void netifd_ubus_interface_notify(struct interface *iface, bool up);
 void netifd_ubus_interface_state_event(struct interface *iface);
+void netifd_ubus_interface_ip_event(struct interface *iface);
 
 #endif
Index: netifd/interface-ip.c
===================================================================
--- netifd.orig/interface-ip.c
+++ netifd/interface-ip.c
@@ -570,6 +570,8 @@ interface_update_proto_addr(struct vlist
 				interface_handle_subnet_route(iface, a_new, true);
 		}
 	}
+
+	netifd_ubus_interface_ip_event(iface);
 }
 
 static bool
@@ -624,6 +626,8 @@ interface_update_proto_route(struct vlis
 		route_new->iface = iface;
 		route_new->enabled = _enabled;
 	}
+
+	netifd_ubus_interface_ip_event(iface);
 }
 
 static void
@@ -1006,6 +1010,8 @@ interface_update_prefix(struct vlist_tre
 	if (node_new && (!prefix_new->iface || !prefix_new->iface->proto_ip.no_delegation))
 		list_add(&prefix_new->head, &prefixes);
 
+	if (tree)
+		netifd_ubus_interface_ip_event(ip->iface);
 }
 
 struct device_prefix*
