Index: netifd/ubus.c
===================================================================
--- netifd.orig/ubus.c
+++ netifd/ubus.c
@@ -450,7 +450,7 @@ interface_ip_dump_address_list(struct in
 }
 
 static void
-interface_ip_dump_route_list(struct interface_ip_settings *ip, bool enabled)
+interface_ip_dump_route_list(struct interface_ip_settings *ip, bool v6, bool enabled)
 {
 	struct device_route *route;
 	int buflen = 128;
@@ -471,6 +471,9 @@ interface_ip_dump_route_list(struct inte
 		else
 			af = AF_INET6;
 
+		if (af != (v6 ? AF_INET6 : AF_INET))
+			continue;
+
 		r = blobmsg_open_table(&b, NULL);
 
 		buf = blobmsg_alloc_string_buffer(&b, "target", buflen);
@@ -702,9 +705,13 @@ netifd_dump_status(struct interface *ifa
 		a = blobmsg_open_array(&b, "ipv6-prefix-assignment");
 		interface_ip_dump_prefix_assignment_list(iface);
 		blobmsg_close_array(&b, a);
-		a = blobmsg_open_array(&b, "route");
-		interface_ip_dump_route_list(&iface->config_ip, true);
-		interface_ip_dump_route_list(&iface->proto_ip, true);
+		a = blobmsg_open_array(&b, "ipv4-route");
+		interface_ip_dump_route_list(&iface->config_ip, false, true);
+		interface_ip_dump_route_list(&iface->proto_ip, false, true);
+		blobmsg_close_array(&b, a);
+		a = blobmsg_open_array(&b, "ipv6-route");
+		interface_ip_dump_route_list(&iface->config_ip, true, true);
+		interface_ip_dump_route_list(&iface->proto_ip, true, true);
 		blobmsg_close_array(&b, a);
 		a = blobmsg_open_array(&b, "dns-server");
 		interface_ip_dump_dns_server_list(&iface->config_ip, true);
@@ -724,9 +731,13 @@ netifd_dump_status(struct interface *ifa
 		interface_ip_dump_address_list(&iface->config_ip, true, false);
 		interface_ip_dump_address_list(&iface->proto_ip, true, false);
 		blobmsg_close_array(&b, a);
-		a = blobmsg_open_array(&b, "route");
-		interface_ip_dump_route_list(&iface->config_ip, false);
-		interface_ip_dump_route_list(&iface->proto_ip, false);
+		a = blobmsg_open_array(&b, "ipv4-route");
+		interface_ip_dump_route_list(&iface->config_ip, false, false);
+		interface_ip_dump_route_list(&iface->proto_ip, false, false);
+		blobmsg_close_array(&b, a);
+		a = blobmsg_open_array(&b, "ipv6-route");
+		interface_ip_dump_route_list(&iface->config_ip, true, false);
+		interface_ip_dump_route_list(&iface->proto_ip, true, false);
 		blobmsg_close_array(&b, a);
 		a = blobmsg_open_array(&b, "dns-server");
 		interface_ip_dump_dns_server_list(&iface->config_ip, false);
