diff -crB clean-0.9.35/src/h_bash.c patched-0.9.35/src/h_bash.c
*** clean-0.9.35/src/h_bash.c	2015-02-16 08:16:19.000000000 +1000
--- patched-0.9.35/src/h_bash.c	2015-10-21 11:48:12.714672378 +1000
***************
*** 64,70 ****
--- 64,74 ----
    retcode = pipe (&subshell_pipe[PARENT_IN]);
    if (retcode == 0)
      {
+ #ifdef __uClinux__
+       subshell_pid = vfork ();
+ #else
        subshell_pid = fork ();
+ #endif
        if (subshell_pid == -1)
  	{
  	  die_with_message (NULL, NULL, g_err_msg[E_SUBSHELL_FAIL]);
diff -crB clean-0.9.35/src/rfc2388.c patched-0.9.35/src/rfc2388.c
*** clean-0.9.35/src/rfc2388.c	2015-02-16 09:01:21.000000000 +1000
--- patched-0.9.35/src/rfc2388.c	2015-10-21 11:48:12.714672378 +1000
***************
*** 216,222 ****
--- 216,226 ----
  
  
  
+ #ifdef __uClinux__
+   pid = vfork ();
+ #else
    pid = fork ();
+ #endif
    if (pid == -1)
      {
        empty_stdin ();
