From: Seth Bollinger <Seth.Bollinger@digi.com>
Date: Tue, 18 Feb 2020 12:23:53 -0600
Subject: build: allow old autoconf version

---
 configure.ac | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/configure.ac b/configure.ac
index 58f44621..10d75952 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1,4 +1,4 @@
-AC_PREREQ([2.63])
+dnl AC_PREREQ([2.63])
 
 dnl-----------------------------------------------------------------------------
 dnl Package and library versioning support
@@ -62,7 +62,7 @@ m4_pattern_forbid([^AX_(COMPILER_FLAGS|COMPILER_FLAGS_(CFLAGS|GIR|LDFLAGS))\b],
   [Unexpanded AX_ macro found. Please install GNU autoconf-archive])
 
 dnl Detect git or release builds
-AX_IS_RELEASE([git-directory])
+dnl AX_IS_RELEASE([git-directory])
 
 dnl Function type cast disabled: used throughout the code especially to
 dnl cast GAsyncReadyCallbacks with the real object type instead of GObject
@@ -72,8 +72,8 @@ dnl All message protocol structs are packed, never complain about it
 DISABLED_WARNINGS="${DISABLED_WARNINGS} -Wno-packed"
 
 dnl Setup compiler checks
-AX_COMPILER_FLAGS()
-AX_COMPILER_FLAGS_CFLAGS(,,,[${DISABLED_WARNINGS}])
+dnl AX_COMPILER_FLAGS()
+dnl AX_COMPILER_FLAGS_CFLAGS(,,,[${DISABLED_WARNINGS}])
 
 dnl Specify gnu89 mode
 if test "$GCC" = "yes"; then
