diff --git a/src/77-mm-usb-device-blacklist.rules b/src/77-mm-usb-device-blacklist.rules
index 6ad4f87c8c..8be8956855 100644
--- a/src/77-mm-usb-device-blacklist.rules
+++ b/src/77-mm-usb-device-blacklist.rules
@@ -203,6 +203,10 @@ ATTRS{idVendor}=="04d8", ATTRS{idProduct}=="f012", ENV{ID_MM_TTY_BLACKLIST}="1"
 # Pycom Pytrack
 ATTRS{idVendor}=="04d8", ATTRS{idProduct}=="f013", ENV{ID_MM_TTY_BLACKLIST}="1"

+# Sierra Boothold
+ATTRS{idVendor}=="1199", ATTRS{idProduct}=="9070", ENV{ID_MM_TTY_BLACKLIST}="1"
+ATTRS{idVendor}=="1199", ATTRS{idProduct}=="9090", ENV{ID_MM_TTY_BLACKLIST}="1"
+
 # All devices from Prusa Research
 ATTRS{idVendor}=="2c99", ENV{ID_MM_TTY_BLACKLIST}="1"

@@ -210,4 +214,7 @@ ATTRS{idVendor}=="2c99", ENV{ID_MM_TTY_BLACKLIST}="1"
 ATTRS{idVendor}=="2548", ATTRS{idProduct}=="1001", ENV{ID_MM_TTY_BLACKLIST}="1"
 ATTRS{idVendor}=="2548", ATTRS{idProduct}=="1002", ENV{ID_MM_TTY_BLACKLIST}="1"

+# Unitac router mode for fw update
+ATTRS{idVendor}=="1076", ATTRS{idProduct}=="9082", ENV{ID_MM_DEVICE_IGNORE}="1"
+
 LABEL="mm_usb_device_blacklist_end"
