commit 734baa3bd8add26ecf40302ee1ed99a246178f56
Author: pchunt <peter.hunt@opengear.com>
Date:   Tue Oct 31 16:32:14 2017 -0600

    OG-5189 OG-3439 Off by 1 sub-parameter length check


Index: ModemManager/src/mm-sms-part-cdma.c
===================================================================
--- ModemManager.orig/src/mm-sms-part-cdma.c
+++ ModemManager/src/mm-sms-part-cdma.c
@@ -846,7 +846,7 @@ read_bearer_data_user_data (MMSmsPart
         gchar *text;
         guint i;
 
-        SUBPARAMETER_SIZE_CHECK (byte_offset + 1 + ((bit_offset + (num_fields * 7)) / 8));
+        SUBPARAMETER_SIZE_CHECK (byte_offset + ((bit_offset + (num_fields * 7)) / 8));
 
         text = g_malloc (num_fields + 1);
         for (i = 0; i < num_fields; i++) {
