--- ModemManager/src/main.c	2018-11-22 11:31:32.261640063 +1000
+++ ModemManager/src/main.c.new	2018-11-22 11:37:28.933640034 +1000
@@ -54,6 +54,24 @@
         exit (0);
     return FALSE;
 }
+static gboolean
+hup_cb (gpointer user_data)
+{
+    GError *err = NULL;
+
+    mm_info ("Reloading log file...");
+    mm_log_shutdown ();
+    if (!mm_log_setup (mm_context_get_log_level (),
+       mm_context_get_log_file (),
+       mm_context_get_log_timestamps (),
+       mm_context_get_log_relative_timestamps (),
+       mm_context_get_debug (),
+       &err)) {
+            g_warning ("Failed to set up logging: %s", err->message);
+            g_error_free (err);
+    }
+    return G_SOURCE_CONTINUE;
+}
 
 #if defined WITH_SYSTEMD_SUSPEND_RESUME
 
@@ -153,6 +171,7 @@
 
     g_unix_signal_add (SIGTERM, quit_cb, NULL);
     g_unix_signal_add (SIGINT, quit_cb, NULL);
+    g_unix_signal_add (SIGHUP, hup_cb, NULL);
 
     mm_info ("ModemManager (version " MM_DIST_VERSION ") starting in %s bus...",
              mm_context_get_test_session () ? "session" : "system");
--- ModemManager/src/mm-log.c    2018-11-13 15:22:45.286516047 +1000
+++ ModemManager/src/mm-log.c.new    2018-11-22 11:37:28.933640034 +1000
@@ -350,6 +350,8 @@
 {
     if (logfd < 0)
         closelog ();
-    else
+    else {
         close (logfd);
+        logfd = -1;
+    }
 }
