diff --git a/plugins/pantech/mm-plugin-pantech.c b/plugins/pantech/mm-plugin-pantech.c
index 788a45d..a82a815 100644
--- a/plugins/pantech/mm-plugin-pantech.c
+++ b/plugins/pantech/mm-plugin-pantech.c
@@ -133,7 +133,7 @@ G_MODULE_EXPORT MMPlugin *
 mm_plugin_create (void)
 {
     static const gchar *subsystems[] = { "tty", "net", "usb", NULL };
-    static const guint16 vendor_ids[] = { 0x106c, 0 };
+    static const guint16 vendor_ids[] = { 0x106c, 0x10a9, 0 };
 
     return MM_PLUGIN (
         g_object_new (MM_TYPE_PLUGIN_PANTECH,
