--- ModemManager-0.7.990/src/mm-log.c.orig	2013-05-26 20:52:16.391045568 +1000
+++ ModemManager-0.7.990/src/mm-log.c	2013-05-26 20:52:26.391062394 +1000
@@ -218,7 +218,7 @@
     g_get_current_time (&rel_start);
 
     if (log_file == NULL)
-        openlog (G_LOG_DOMAIN, LOG_CONS | LOG_PID | LOG_PERROR, LOG_DAEMON);
+        openlog (G_LOG_DOMAIN, LOG_CONS | LOG_PID, LOG_DAEMON);
     else {
         logfd = open (log_file,
                       O_CREAT | O_APPEND | O_WRONLY,
