diff -ru ModemManager/src/mm-bearer-qmi.c ModemManager-patched/src/mm-bearer-qmi.c
--- ModemManager/src/mm-bearer-qmi.c	2017-07-06 12:17:52.954453400 +1000
+++ ModemManager/src/mm-bearer-qmi.c	2017-07-06 12:17:55.434453440 +1000
@@ -957,9 +957,11 @@
          * to request. If the LLP is raw-ip, we force Static IP, because not
          * all DHCP clients support the raw-ip interfaces; otherwise default
          * to DHCP as always. */
+#if 0
         if (mm_port_qmi_llp_is_raw_ip (ctx->qmi))
             ctx->ip_method = MM_BEARER_IP_METHOD_STATIC;
         else
+#endif
             ctx->ip_method = MM_BEARER_IP_METHOD_DHCP;
 
         mm_dbg ("Defaulting to use %s IP method", mm_bearer_ip_method_get_string (ctx->ip_method));
