diff --git a/src/mm-broadband-modem.c b/src/mm-broadband-modem.c
index c9b438a..9b3fdbe 100644
--- a/src/mm-broadband-modem.c
+++ b/src/mm-broadband-modem.c
@@ -8074,7 +8074,8 @@ disabling_context_complete_and_free (DisablingContext *ctx)
         mm_iface_modem_update_state (MM_IFACE_MODEM (ctx->self),
                                      MM_MODEM_STATE_DISABLED,
                                      MM_MODEM_STATE_CHANGE_REASON_USER_REQUESTED);
-    else if (ctx->previous_state != MM_MODEM_STATE_DISABLED) {
+    else if (ctx->previous_state != MM_MODEM_STATE_DISABLED &&
+             ctx->previous_state != MM_MODEM_STATE_UNKNOWN) {
         /* Fallback to previous state */
         mm_iface_modem_update_state (MM_IFACE_MODEM (ctx->self),
                                      ctx->previous_state,
@@ -8428,7 +8429,8 @@ enabling_context_complete_and_free (EnablingContext *ctx)
         mm_iface_modem_update_state (MM_IFACE_MODEM (ctx->self),
                                      MM_MODEM_STATE_ENABLED,
                                      MM_MODEM_STATE_CHANGE_REASON_USER_REQUESTED);
-    else if (ctx->previous_state != MM_MODEM_STATE_ENABLED) {
+    else if (ctx->previous_state != MM_MODEM_STATE_ENABLED &&
+             ctx->previous_state != MM_MODEM_STATE_UNKNOWN) {
         /* Fallback to previous state */
         mm_iface_modem_update_state (MM_IFACE_MODEM (ctx->self),
                                      ctx->previous_state,
