diff -Nurd ModemManager.orig/src/mm-bearer-qmi.c ModemManager/src/mm-bearer-qmi.c
--- ModemManager.orig/src/mm-bearer-qmi.c	2019-05-29 11:08:19.348305221 +1000
+++ ModemManager/src/mm-bearer-qmi.c	2019-05-29 11:08:19.416305556 +1000
@@ -1814,13 +1814,11 @@
             ctx->error_ipv4 = error;
         else
             ctx->error_ipv6 = error;
-    } else {
-        /* Clear internal status */
-        reset_bearer_connection (self,
-                                 ctx->running_ipv4,
-                                 ctx->running_ipv6);
     }
 
+	/* Clear internal status */
+	reset_bearer_connection (self, TRUE, TRUE);
+
     if (output)
         qmi_message_wds_stop_network_output_unref (output);
 
