diff -Nurd ModemManager.orig/src/mm-bearer-qmi.c ModemManager/src/mm-bearer-qmi.c
--- ModemManager.orig/src/mm-bearer-qmi.c	2019-05-29 11:08:19.416305556 +1000
+++ ModemManager/src/mm-bearer-qmi.c	2019-05-29 11:08:19.436305652 +1000
@@ -531,28 +531,51 @@
             if (g_error_matches (error,
                                  QMI_PROTOCOL_ERROR,
                                  QMI_PROTOCOL_ERROR_CALL_FAILED)) {
-                QmiWdsCallEndReason cer;
-                QmiWdsVerboseCallEndReasonType verbose_cer_type;
-                gint16 verbose_cer_reason;
+                QmiWdsCallEndReason cer = 0;
+                QmiWdsVerboseCallEndReasonType verbose_cer_type = 0;
+                gint16 verbose_cer_reason = 0;
+                const char *cer_str = NULL;
+                const char *verbose_cer_type_str = NULL;
+                const char *verbose_cer_reason_str = NULL;
+                char *str;
+
 
                 if (qmi_message_wds_start_network_output_get_call_end_reason (
                         output,
                         &cer,
-                        NULL))
+                        NULL)) {
+                    cer_str = qmi_wds_call_end_reason_get_string (cer);
                     mm_info ("call end reason (%u): '%s'",
                              cer,
-                             qmi_wds_call_end_reason_get_string (cer));
+                             cer_str);
+				}
 
                 if (qmi_message_wds_start_network_output_get_verbose_call_end_reason (
                         output,
                         &verbose_cer_type,
                         &verbose_cer_reason,
-                        NULL))
+                        NULL)) {
+                    verbose_cer_type_str = qmi_wds_verbose_call_end_reason_type_get_string (verbose_cer_type);
+                    verbose_cer_reason_str = qmi_wds_verbose_call_end_reason_get_string (verbose_cer_type, verbose_cer_reason);
                     mm_info ("verbose call end reason (%u,%d): [%s] %s",
                              verbose_cer_type,
                              verbose_cer_reason,
-                             qmi_wds_verbose_call_end_reason_type_get_string (verbose_cer_type),
-                             qmi_wds_verbose_call_end_reason_get_string (verbose_cer_type, verbose_cer_reason));
+							 verbose_cer_type_str,
+							 verbose_cer_reason_str);
+				}
+
+				if (asprintf(&str, "mm-call-end-notify.sh \"%s\" \"%s\" \"%s %s %s (%u,%u,%u)\"",
+					mm_base_bearer_get_path(&ctx->self->parent),
+					ctx->running_ipv6 ? "ipv6" : "ipv4",
+					cer_str ?: "",
+					verbose_cer_type_str ?: "",
+					verbose_cer_reason_str ?: "",
+					cer,
+					verbose_cer_type,
+					verbose_cer_reason) != -1) {
+						system(str);
+						free(str);
+				}
             }
         }
     }
@@ -964,28 +987,48 @@
         if (connection_status == QMI_WDS_CONNECTION_STATUS_DISCONNECTED &&
             bearer_status != MM_BEARER_STATUS_DISCONNECTED &&
             bearer_status != MM_BEARER_STATUS_DISCONNECTING) {
-            QmiWdsCallEndReason cer;
-            QmiWdsVerboseCallEndReasonType verbose_cer_type;
-            gint16 verbose_cer_reason;
+            QmiWdsCallEndReason cer = 0;
+            QmiWdsVerboseCallEndReasonType verbose_cer_type = 0;
+            gint16 verbose_cer_reason = 0;
+            const char *cer_str = NULL;
+            const char *verbose_cer_type_str = NULL;
+            const char *verbose_cer_reason_str = NULL;
+            char *str;
 
             if (qmi_indication_wds_packet_service_status_output_get_call_end_reason (
                     output,
                     &cer,
-                    NULL))
+                    NULL)) {
                 mm_info ("bearer call end reason (%u): '%s'",
                          cer,
-                         qmi_wds_call_end_reason_get_string (cer));
+						 cer_str);
+			}
 
             if (qmi_indication_wds_packet_service_status_output_get_verbose_call_end_reason (
                     output,
                     &verbose_cer_type,
                     &verbose_cer_reason,
-                    NULL))
+                    NULL)) {
+                verbose_cer_type_str = qmi_wds_verbose_call_end_reason_type_get_string (verbose_cer_type);
+                verbose_cer_reason_str = qmi_wds_verbose_call_end_reason_get_string (verbose_cer_type, verbose_cer_reason);
                 mm_info ("bearer verbose call end reason (%u,%d): [%s] %s",
                          verbose_cer_type,
                          verbose_cer_reason,
-                         qmi_wds_verbose_call_end_reason_type_get_string (verbose_cer_type),
-                         qmi_wds_verbose_call_end_reason_get_string (verbose_cer_type, verbose_cer_reason));
+						 verbose_cer_type_str,
+						 verbose_cer_reason_str);
+			}
+
+			if (asprintf(&str, "mm-call-end-notify.sh \"%s\" \"\" \"%s %s %s (%u,%u,%u)\"",
+				mm_base_bearer_get_path(&self->parent),
+				cer_str ?: "",
+				verbose_cer_type_str ?: "",
+				verbose_cer_reason_str ?: "",
+				cer,
+				verbose_cer_type,
+				verbose_cer_reason) != -1) {
+					system(str);
+					free(str);
+			}
 
             mm_base_bearer_report_connection_status (MM_BASE_BEARER (self), MM_BEARER_CONNECTION_STATUS_DISCONNECTED);
         }
