diff --git a/src/mm-bearer-qmi.c b/src/mm-bearer-qmi.c
index db8e0ac..46a4045 100644
--- a/src/mm-bearer-qmi.c
+++ b/src/mm-bearer-qmi.c
@@ -1164,13 +1164,11 @@ stop_network_ready (QmiClientWds *client,
             ctx->error_ipv4 = error;
         else
             ctx->error_ipv6 = error;
-    } else {
-        /* Clear internal status */
-        reset_bearer_connection (ctx->self,
-                                 ctx->running_ipv4,
-                                 ctx->running_ipv6);
     }
 
+    /* Clear internal status */
+    reset_bearer_connection (ctx->self, TRUE, TRUE);
+
     if (output)
         qmi_message_wds_stop_network_output_unref (output);
 
