From 86f2ac1e660af39bc51d5521666ad5a2b7df2345 Mon Sep 17 00:00:00 2001
From: Aleksander Morgado <aleksander@aleksander.es>
Date: Tue, 18 Oct 2016 10:28:50 +0200
Subject: [PATCH] broadband-modem-qmi: don't use PIN2 lock state if unknown

---
 src/mm-broadband-modem-qmi.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/src/mm-broadband-modem-qmi.c b/src/mm-broadband-modem-qmi.c
index bc13925..b4f7b56 100644
--- a/src/mm-broadband-modem-qmi.c
+++ b/src/mm-broadband-modem-qmi.c
@@ -1792,8 +1792,14 @@ dms_uim_get_pin_status_ready (QmiClientDms *client,
             &current_status,
             NULL, /* verify_retries_left */
             NULL, /* unblock_retries_left */
-            NULL))
-        lock = mm_modem_lock_from_qmi_uim_pin_status (current_status, FALSE);
+            NULL)) {
+        MMModemLock lock2;
+
+        /* We only use the PIN2 status if it isn't unknown */
+        lock2 = mm_modem_lock_from_qmi_uim_pin_status (current_status, FALSE);
+        if (lock2 != MM_MODEM_LOCK_UNKNOWN)
+            lock = lock2;
+    }
 
     /* We're done! */
     g_simple_async_result_set_op_res_gpointer (ctx->result, GUINT_TO_POINTER (lock), NULL);
-- 
2.10.0

