diff --git a/plugins/cinterion/mm-plugin-cinterion.c b/plugins/cinterion/mm-plugin-cinterion.c
index 18c5319..40053a9 100644
--- a/plugins/cinterion/mm-plugin-cinterion.c
+++ b/plugins/cinterion/mm-plugin-cinterion.c
@@ -176,6 +176,8 @@ grab_port (MMPlugin *self,
                 mm_port_probe_get_port_subsys (probe),
                 mm_port_probe_get_port_name (probe));
         pflags = MM_PORT_SERIAL_AT_FLAG_PPP;
+	if (g_list_length(mm_device_peek_port_probe_list(self)) < 2)
+		pflags |= MM_PORT_SERIAL_AT_FLAG_SECONDARY;
     } else if (g_udev_device_get_property_as_boolean (mm_port_probe_peek_port (probe),
                                                       "ID_MM_CINTERION_PORT_TYPE_GPS")) {
         mm_dbg ("(%s/%s)' Port flagged as GPS",
