diff -Nurd ModemManager.orig/src/mm-iface-modem.c ModemManager/src/mm-iface-modem.c
--- ModemManager.orig/src/mm-iface-modem.c	2019-05-29 11:08:19.096303978 +1000
+++ ModemManager/src/mm-iface-modem.c	2019-05-29 11:08:19.128304133 +1000
@@ -748,11 +748,11 @@
         return;
     }
 
+#if ! defined WITH_AT_COMMAND_VIA_DBUS
     if (abort_invocation_if_state_not_reached (ctx->self, ctx->invocation, MM_MODEM_STATE_LOCKED)) {
         handle_command_context_free (ctx);
         return;
     }
 
-#if ! defined WITH_AT_COMMAND_VIA_DBUS
     /* If we are not in Debug mode, report an error */
     if (!mm_context_get_debug ()) {
         g_dbus_method_invocation_return_error (ctx->invocation,
