diff -Nurd ModemManager.orig/plugins/novatel/mm-sim-novatel-lte.c ModemManager/plugins/novatel/mm-sim-novatel-lte.c
--- ModemManager.orig/plugins/novatel/mm-sim-novatel-lte.c	2019-05-27 15:37:56.000000000 +1000
+++ ModemManager/plugins/novatel/mm-sim-novatel-lte.c	2019-05-29 11:08:19.516306050 +1000
@@ -66,8 +66,8 @@
     str = mm_strip_tag (response, "+CRSM:");
 
     /* With or without quotes... */
-    if (sscanf (str, "%d,%d,\"%18c\"", &sw1, &sw2, (char *) &buf) != 3 &&
-        sscanf (str, "%d,%d,%18c", &sw1, &sw2, (char *) &buf) != 3) {
+    if (sscanf (str, "%d,%d,\"%18s\"", &sw1, &sw2, (char *) &buf) != 3 &&
+        sscanf (str, "%d,%d,%18s", &sw1, &sw2, (char *) &buf) != 3) {
         g_task_return_new_error (task,
                                  MM_CORE_ERROR,
                                  MM_CORE_ERROR_FAILED,
