--- ModemManager/src/mm-iface-modem.c	2018-11-20 13:07:04.158319614 +1000
+++ ModemManager/src/mm-iface-modem.c.new	2018-11-20 13:23:52.122319532 +1000
@@ -4983,6 +4983,15 @@
                           "handle-factory-reset",
                           G_CALLBACK (handle_factory_reset),
                           self);
+        /* Allow AT commands in the FAILED state as this operation does not
+         * necessarily depend on the presence of a SIM. This is important for
+         * modems that support multiple SIM slots and selection is performed
+         * with an AT command. */
+        g_signal_connect (ctx->skeleton,
+                          "handle-command",
+                          G_CALLBACK (handle_command),
+                          self);
+
 
         if (ctx->fatal_error) {
             if (g_error_matches (ctx->fatal_error,
@@ -5006,10 +5015,6 @@
                               G_CALLBACK (handle_create_bearer),
                               self);
             g_signal_connect (ctx->skeleton,
-                              "handle-command",
-                              G_CALLBACK (handle_command),
-                              self);
-            g_signal_connect (ctx->skeleton,
                               "handle-delete-bearer",
                               G_CALLBACK (handle_delete_bearer),
                               self);
