diff -Nurd ModemManager.orig/src/mm-iface-modem-3gpp-ussd.h ModemManager/src/mm-iface-modem-3gpp-ussd.h
--- ModemManager.orig/src/mm-iface-modem-3gpp-ussd.h	2019-05-27 15:37:56.000000000 +1000
+++ ModemManager/src/mm-iface-modem-3gpp-ussd.h	2019-05-29 11:08:19.148304234 +1000
@@ -29,8 +29,8 @@
 #define MM_IFACE_MODEM_3GPP_USSD_GET_INTERFACE(obj) (G_TYPE_INSTANCE_GET_INTERFACE ((obj), MM_TYPE_IFACE_MODEM_3GPP_USSD, MMIfaceModem3gppUssd))
 
 /* CBS data coding scheme - 3GPP TS 23.038 */
-#define MM_MODEM_GSM_USSD_SCHEME_7BIT 0b00001111
-#define MM_MODEM_GSM_USSD_SCHEME_UCS2 0b01001000
+#define MM_MODEM_GSM_USSD_SCHEME_7BIT 0x0f //0b00001111
+#define MM_MODEM_GSM_USSD_SCHEME_UCS2 0x48 //0b01001000
 
 #define MM_IFACE_MODEM_3GPP_USSD_DBUS_SKELETON "iface-modem-3gpp-ussd-dbus-skeleton"
 
