diff --git a/plugins/novatel/77-mm-novatel-u620l-enterprise-mode.rules b/plugins/novatel/77-mm-novatel-u620l-enterprise-mode.rules
new file mode 100644
index 0000000..4179c2a
--- /dev/null
+++ b/plugins/novatel/77-mm-novatel-u620l-enterprise-mode.rules
@@ -0,0 +1 @@
+ATTRS{idVendor}=="1410", ATTRS{idProduct}=="9020", ATTR{bConfigurationValue}="4"
--- ModemManager/plugins/Makefile.am	2018-11-13 15:22:44.838516047 +1000
+++ ModemManager/plugins/Makefile.am.new	2018-11-26 16:05:44.956943765 +1000
@@ -246,6 +246,8 @@
 NOVATEL_COMMON_COMPILER_FLAGS = -I$(top_srcdir)/plugins/novatel
 NOVATEL_COMMON_LIBADD_FLAGS   = $(builddir)/libmm-utils-novatel.la
 
+dist_udevrules_DATA += novatel/77-mm-novatel-u620l-enterprise-mode.rules
+
 ################################################################################
 # common xmm support
 ################################################################################
