diff -Nurd ModemManager.orig/plugins/pantech/mm-plugin-pantech.c ModemManager/plugins/pantech/mm-plugin-pantech.c
--- ModemManager.orig/plugins/pantech/mm-plugin-pantech.c	2019-05-27 15:37:56.000000000 +1000
+++ ModemManager/plugins/pantech/mm-plugin-pantech.c	2019-05-29 11:08:19.536306146 +1000
@@ -131,7 +131,7 @@
 mm_plugin_create (void)
 {
     static const gchar *subsystems[] = { "tty", "net", "usb", NULL };
-    static const guint16 vendor_ids[] = { 0x106c, 0 };
+    static const guint16 vendor_ids[] = { 0x106c, 0x10a9, 0 };
 
     return MM_PLUGIN (
         g_object_new (MM_TYPE_PLUGIN_PANTECH,
