diff -Nurd ModemManager.orig/src/mm-broadband-modem.c ModemManager/src/mm-broadband-modem.c
--- ModemManager.orig/src/mm-broadband-modem.c	2019-05-27 15:37:56.000000000 +1000
+++ ModemManager/src/mm-broadband-modem.c	2019-05-29 11:08:19.260304787 +1000
@@ -9667,7 +9667,8 @@
         mm_iface_modem_update_state (MM_IFACE_MODEM (ctx->self),
                                      MM_MODEM_STATE_DISABLED,
                                      MM_MODEM_STATE_CHANGE_REASON_USER_REQUESTED);
-    else if (ctx->previous_state != MM_MODEM_STATE_DISABLED) {
+    else if (ctx->previous_state != MM_MODEM_STATE_DISABLED &&
+             ctx->previous_state != MM_MODEM_STATE_UNKNOWN) {
         /* Fallback to previous state */
         mm_iface_modem_update_state (MM_IFACE_MODEM (ctx->self),
                                      ctx->previous_state,
@@ -10027,7 +10028,8 @@
         mm_iface_modem_update_state (MM_IFACE_MODEM (ctx->self),
                                      MM_MODEM_STATE_ENABLED,
                                      MM_MODEM_STATE_CHANGE_REASON_USER_REQUESTED);
-    else if (ctx->previous_state != MM_MODEM_STATE_ENABLED) {
+    else if (ctx->previous_state != MM_MODEM_STATE_ENABLED &&
+             ctx->previous_state != MM_MODEM_STATE_UNKNOWN) {
         /* Fallback to previous state */
         mm_iface_modem_update_state (MM_IFACE_MODEM (ctx->self),
                                      ctx->previous_state,
