diff -Nurd ModemManager.orig/src/mm-log.c ModemManager/src/mm-log.c
--- ModemManager.orig/src/mm-log.c	2019-05-29 11:08:19.476305853 +1000
+++ ModemManager/src/mm-log.c	2019-05-29 11:08:19.496305950 +1000
@@ -308,7 +308,7 @@
     } else
 #endif
     if (log_file == NULL) {
-        openlog (G_LOG_DOMAIN, LOG_CONS | LOG_PID | LOG_PERROR, LOG_DAEMON);
+        openlog (G_LOG_DOMAIN, LOG_CONS | LOG_PID, LOG_DAEMON);
         log_backend = log_backend_syslog;
     } else {
         logfd = open (log_file,
