diff --git a/src/mm-iface-modem-location.c b/src/mm-iface-modem-location.c
index 32f9a73..bfcfb43 100644
--- a/src/mm-iface-modem-location.c
+++ b/src/mm-iface-modem-location.c
@@ -1372,7 +1372,7 @@ mm_iface_modem_location_initialize (MMIfaceModemLocation *self,
         /* Set all initial property defaults */
         mm_gdbus_modem_location_set_capabilities (skeleton, MM_MODEM_LOCATION_SOURCE_NONE);
         mm_gdbus_modem_location_set_enabled (skeleton, MM_MODEM_LOCATION_SOURCE_NONE);
-        mm_gdbus_modem_location_set_signals_location (skeleton, FALSE);
+        mm_gdbus_modem_location_set_signals_location (skeleton, TRUE);
         mm_gdbus_modem_location_set_location (skeleton,
                                               build_location_dictionary (NULL, NULL, NULL, NULL, NULL));
 
