diff -Nurd ModemManager.orig/src/mm-bearer-qmi.c ModemManager/src/mm-bearer-qmi.c
--- ModemManager.orig/src/mm-bearer-qmi.c	2019-05-27 15:37:56.000000000 +1000
+++ ModemManager/src/mm-bearer-qmi.c	2019-05-29 11:08:19.348305221 +1000
@@ -1242,9 +1242,11 @@
          * to request. If the LLP is raw-ip, we force Static IP, because not
          * all DHCP clients support the raw-ip interfaces; otherwise default
          * to DHCP as always. */
+#if 0
         if (mm_port_qmi_llp_is_raw_ip (ctx->qmi))
             ctx->ip_method = MM_BEARER_IP_METHOD_STATIC;
         else
+#endif
             ctx->ip_method = MM_BEARER_IP_METHOD_DHCP;
 
         mm_dbg ("Defaulting to use %s IP method", mm_bearer_ip_method_get_string (ctx->ip_method));
