diff -Nurd ModemManager.orig/plugins/telit/77-mm-telit-port-types.rules ModemManager/plugins/telit/77-mm-telit-port-types.rules
--- ModemManager.orig/plugins/telit/77-mm-telit-port-types.rules	2019-05-27 15:37:56.000000000 +1000
+++ ModemManager/plugins/telit/77-mm-telit-port-types.rules	2019-05-29 11:08:19.392305438 +1000
@@ -33,4 +33,8 @@
 # CE910-DUAL
 ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="1011", ENV{.MM_USBIFNUM}=="01", ENV{ID_MM_PORT_TYPE_AT_PRIMARY}="1"
 
+# LM940
+ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="1040", ENV{ID_MM_TELIT_TAGGED}="1", ENV{ID_MM_TELIT_PORTS_TAGGED}="1"
+ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="1041", ENV{ID_MM_TELIT_TAGGED}="1", ENV{ID_MM_TELIT_PORTS_TAGGED}="1"
+
 LABEL="mm_telit_port_types_end"
