--- ModemManager/plugins/telit/77-mm-telit-port-types.rules	2018-11-20 16:38:05.118318590 +1000
+++ ModemManager/plugins/telit/77-mm-telit-port-types.rules.new	2018-11-20 16:55:56.854318504 +1000
@@ -37,4 +37,7 @@
 ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="1040", ENV{ID_MM_TELIT_TAGGED}="1", ENV{ID_MM_TELIT_PORTS_TAGGED}="1"
 ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="1041", ENV{ID_MM_TELIT_TAGGED}="1", ENV{ID_MM_TELIT_PORTS_TAGGED}="1"
 
+# LE910 V2
+ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="0032", ENV{ID_MM_TELIT_TAGGED}="1", ENV{ID_MM_TELIT_PORTS_TAGGED}="1"
+
 LABEL="mm_telit_port_types_end"
