--- ModemManager/plugins/telit/77-mm-telit-port-types.rules	2018-11-21 08:15:01.618314046 +1000
+++ ModemManager/plugins/telit/77-mm-telit-port-types.rules.new	2018-11-21 08:20:39.034314019 +1000
@@ -41,4 +41,10 @@
 ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="0036", ENV{.MM_USBIFNUM}=="00", RUN += "/bin/telit-mbim.sh"
 ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="0032", ENV{ID_MM_TELIT_TAGGED}="1", ENV{ID_MM_TELIT_PORTS_TAGGED}="1"
 
+# LE910C1
+ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="1201", ENV{.MM_USBIFNUM}=="05", ENV{ID_MM_TELIT_TAGGED}="1", RUN += "/bin/telit-mbim.sh"
+ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="1204", ENV{.MM_USBIFNUM}=="05", ENV{ID_MM_TELIT_PORT_TYPE_MODEM}="1"
+ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="1204", ENV{.MM_USBIFNUM}=="06", ENV{ID_MM_TELIT_PORT_TYPE_AUX}="1"
+ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="1204", ENV{ID_MM_TELIT_TAGGED}="1"
+
 LABEL="mm_telit_port_types_end"
