diff -Nurd ModemManager.orig/plugins/telit/mm-broadband-modem-telit.c ModemManager/plugins/telit/mm-broadband-modem-telit.c
--- ModemManager.orig/plugins/telit/mm-broadband-modem-telit.c	2019-05-27 15:37:56.000000000 +1000
+++ ModemManager/plugins/telit/mm-broadband-modem-telit.c	2019-05-29 11:08:19.372305337 +1000
@@ -830,6 +830,9 @@
         case 3:
             *result = g_variant_new_uint32 (MM_MODEM_ACCESS_TECHNOLOGY_EVDOA);
             return TRUE;
+        case 4:
+            *result = g_variant_new_uint32 (MM_MODEM_ACCESS_TECHNOLOGY_LTE);
+            return TRUE;
         default:
             break;
         }
