diff -Nurd ModemManager.orig/src/mm-plugin-manager.c ModemManager/src/mm-plugin-manager.c
--- ModemManager.orig/src/mm-plugin-manager.c	2019-05-27 15:37:56.000000000 +1000
+++ ModemManager/src/mm-plugin-manager.c	2019-05-29 11:08:19.620306563 +1000
@@ -671,11 +671,11 @@
 /* Device context */
 
 /* Time to wait for ports to appear before starting to probe the first one */
-#define MIN_WAIT_TIME_MSECS 1500
+#define MIN_WAIT_TIME_MSECS 7500
 
 /* Time to wait for other ports to appear once the first port is exposed
  * (needs to be > MIN_WAIT_TIME_MSECS!!) */
-#define MIN_PROBING_TIME_MSECS 2500
+#define MIN_PROBING_TIME_MSECS 10000
 
 /* The wait time we define must always be less than the probing time */
 G_STATIC_ASSERT (MIN_WAIT_TIME_MSECS < MIN_PROBING_TIME_MSECS);
