diff -Nurd ModemManager.orig/m4/vapigen.m4 ModemManager/m4/vapigen.m4
--- ModemManager.orig/m4/vapigen.m4	2019-05-27 15:37:56.000000000 +1000
+++ ModemManager/m4/vapigen.m4	2019-05-29 11:08:19.288304925 +1000
@@ -23,7 +23,7 @@
 # See http://live.gnome.org/Vala/UpstreamGuide for detailed documentation
 AC_DEFUN([VAPIGEN_CHECK],
 [
-  AS_IF([test "x$3" != "xyes"], [
+  AS_IF([test "x$3" != "xyes"],:;[
       m4_provide_if([GOBJECT_INTROSPECTION_CHECK], [], [
           m4_provide_if([GOBJECT_INTROSPECTION_REQUIRE], [], [
               AC_MSG_ERROR([[You must call GOBJECT_INTROSPECTION_CHECK or GOBJECT_INTROSPECTION_REQUIRE before using VAPIGEN_CHECK unless using the FOUND_INTROSPECTION argument is "yes"]])
