--- ModemManager-1.2.0/src/mm-iface-modem.c.orig	2014-02-11 11:21:25.369612168 +1000
+++ ModemManager-1.2.0/src/mm-iface-modem.c	2014-02-11 11:21:32.499449674 +1000
@@ -699,7 +699,7 @@
     }
 
     /* If we are not in Debug mode, report an error */
-    if (!mm_context_get_debug ()) {
+    if (0 && !mm_context_get_debug ()) {
         g_dbus_method_invocation_return_error (ctx->invocation,
                                                MM_CORE_ERROR,
                                                MM_CORE_ERROR_UNAUTHORIZED,
