--- ModemManager-0.7.990/src/mm-iface-modem-3gpp-ussd.h.orig	2013-05-15 13:09:15.560428260 +1000
+++ ModemManager-0.7.990/src/mm-iface-modem-3gpp-ussd.h	2013-05-15 13:09:17.560400535 +1000
@@ -31,8 +31,8 @@
 #define MM_IFACE_MODEM_3GPP_USSD_GET_INTERFACE(obj) (G_TYPE_INSTANCE_GET_INTERFACE ((obj), MM_TYPE_IFACE_MODEM_3GPP_USSD, MMIfaceModem3gppUssd))
 
 /* CBS data coding scheme - 3GPP TS 23.038 */
-#define MM_MODEM_GSM_USSD_SCHEME_7BIT 0b00001111
-#define MM_MODEM_GSM_USSD_SCHEME_UCS2 0b01001000
+#define MM_MODEM_GSM_USSD_SCHEME_7BIT 0x0f //0b00001111
+#define MM_MODEM_GSM_USSD_SCHEME_UCS2 0x48 //0b01001000
 
 #define MM_IFACE_MODEM_3GPP_USSD_DBUS_SKELETON "iface-modem-3gpp-ussd-dbus-skeleton"
 
