diff -Nurd ModemManager.orig/plugins/Makefile.am ModemManager/plugins/Makefile.am
--- ModemManager.orig/plugins/Makefile.am	2019-05-27 15:37:56.000000000 +1000
+++ ModemManager/plugins/Makefile.am	2019-05-29 11:08:19.556306247 +1000
@@ -250,6 +250,8 @@
 NOVATEL_COMMON_COMPILER_FLAGS = -I$(top_srcdir)/plugins/novatel
 NOVATEL_COMMON_LIBADD_FLAGS   = $(builddir)/libmm-utils-novatel.la
 
+dist_udevrules_DATA += novatel/77-mm-novatel-u620l-enterprise-mode.rules
+
 ################################################################################
 # common xmm support
 ################################################################################
diff -Nurd ModemManager.orig/plugins/novatel/77-mm-novatel-u620l-enterprise-mode.rules ModemManager/plugins/novatel/77-mm-novatel-u620l-enterprise-mode.rules
--- ModemManager.orig/plugins/novatel/77-mm-novatel-u620l-enterprise-mode.rules	1970-01-01 10:00:00.000000000 +1000
+++ ModemManager/plugins/novatel/77-mm-novatel-u620l-enterprise-mode.rules	2019-05-29 11:08:19.556306247 +1000
@@ -0,0 +1 @@
+ATTRS{idVendor}=="1410", ATTRS{idProduct}=="9020", ATTR{bConfigurationValue}="4"
