diff --git a/cli/mmcli-bearer.c b/cli/mmcli-bearer.c
index cd9ecca..ebea037 100644
--- a/cli/mmcli-bearer.c
+++ b/cli/mmcli-bearer.c
@@ -172,6 +172,7 @@ print_bearer_info (MMBearer *bearer)
                  "                     |     IP type: '%s'\n"
                  "                     |        user: '%s'\n"
                  "                     |    password: '%s'\n"
+                 "                     |  allow auth: '%s'\n"
                  "                     |      number: '%s'\n"
                  "                     | Rm protocol: '%s'\n",
                  VALIDATE_NONE (mm_bearer_properties_get_apn (properties)),
@@ -179,6 +180,7 @@ print_bearer_info (MMBearer *bearer)
                  VALIDATE_UNKNOWN (ip_family_str),
                  VALIDATE_NONE (mm_bearer_properties_get_user (properties)),
                  VALIDATE_NONE (mm_bearer_properties_get_password (properties)),
+                 mm_bearer_allowed_auth_build_string_from_mask(mm_bearer_properties_get_allowed_auth(properties)),
                  VALIDATE_NONE (mm_bearer_properties_get_number (properties)),
                  VALIDATE_UNKNOWN (mm_modem_cdma_rm_protocol_get_string (
                                        mm_bearer_properties_get_rm_protocol (properties))));
