diff -Nurd ModemManager.orig/plugins/telit/77-mm-telit-port-types.rules ModemManager/plugins/telit/77-mm-telit-port-types.rules
--- ModemManager.orig/plugins/telit/77-mm-telit-port-types.rules	2019-05-29 11:08:19.392305438 +1000
+++ ModemManager/plugins/telit/77-mm-telit-port-types.rules	2019-05-29 11:08:19.456305753 +1000
@@ -37,4 +37,7 @@
 ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="1040", ENV{ID_MM_TELIT_TAGGED}="1", ENV{ID_MM_TELIT_PORTS_TAGGED}="1"
 ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="1041", ENV{ID_MM_TELIT_TAGGED}="1", ENV{ID_MM_TELIT_PORTS_TAGGED}="1"
 
+# LE910 V2
+ATTRS{idVendor}=="1bc7", ATTRS{idProduct}=="0032", ENV{ID_MM_TELIT_TAGGED}="1", ENV{ID_MM_TELIT_PORTS_TAGGED}="1"
+
 LABEL="mm_telit_port_types_end"
