Index: libubox/utils.h
===================================================================
--- libubox.orig/utils.h
+++ libubox/utils.h
@@ -105,7 +105,7 @@ static inline uint16_t __u_bswap16(uint1
 	return ((val >> 8) & 0xffu) | ((val & 0xffu) << 8);
 }
 
-#if _GNUC_MIN_VER(4, 2)
+#if _GNUC_MIN_VER(4, 3)
 #define __u_bswap32(x) __builtin_bswap32(x)
 #define __u_bswap64(x) __builtin_bswap64(x)
 #else
