Index: iproute2-3.15.0/ip/iplink.c
===================================================================
--- iproute2-3.15.0.orig/ip/iplink.c	2014-06-11 02:39:14.000000000 +1000
+++ iproute2-3.15.0/ip/iplink.c	2014-06-17 16:40:31.276734502 +1000
@@ -33,7 +33,7 @@
 #include "utils.h"
 #include "ip_common.h"
 
-#define IPLINK_IOCTL_COMPAT	1
+#define IPLINK_IOCTL_COMPAT	0
 #ifndef LIBDIR
 #define LIBDIR "/usr/lib"
 #endif
