diff -Naur iproute2-5.5.0.orig/tc/Makefile iproute2-5.5.0/tc/Makefile
--- iproute2-5.5.0.orig/tc/Makefile	2020-01-27 23:53:09.000000000 +1000
+++ iproute2-5.5.0/tc/Makefile	2020-03-09 16:39:01.084627194 +1000
@@ -1,84 +1,91 @@
 # SPDX-License-Identifier: GPL-2.0
 TCOBJ= tc.o tc_qdisc.o tc_class.o tc_filter.o tc_util.o tc_monitor.o \
-       tc_exec.o m_police.o m_estimator.o m_action.o m_ematch.o \
-       emp_ematch.tab.o emp_ematch.lex.o
+       tc_exec.o m_police.o m_estimator.o m_action.o
+
+include $(LINUX_CONFIG)
+include $(CONFIG_CONFIG)
+-include $(MODULES_CONFIG)
+ifdef CONFIG_NET_EMATCH
+TCOBJ += m_ematch.o emp_ematch.tab.o emp_ematch.lex.o
+endif
 
 include ../config.mk
 
 SHARED_LIBS ?= y
 
-TCMODULES :=
-TCMODULES += q_fifo.o
-TCMODULES += q_sfq.o
-TCMODULES += q_red.o
-TCMODULES += q_prio.o
-TCMODULES += q_skbprio.o
-TCMODULES += q_tbf.o
-TCMODULES += q_cbq.o
-TCMODULES += q_rr.o
-TCMODULES += q_multiq.o
-TCMODULES += q_netem.o
-TCMODULES += q_choke.o
-TCMODULES += q_sfb.o
-TCMODULES += f_rsvp.o
-TCMODULES += f_u32.o
-TCMODULES += f_route.o
-TCMODULES += f_fw.o
-TCMODULES += f_basic.o
-TCMODULES += f_bpf.o
-TCMODULES += f_flow.o
-TCMODULES += f_cgroup.o
-TCMODULES += f_flower.o
-TCMODULES += q_dsmark.o
-TCMODULES += q_gred.o
-TCMODULES += f_tcindex.o
-TCMODULES += q_ingress.o
-TCMODULES += q_hfsc.o
-TCMODULES += q_htb.o
-TCMODULES += q_drr.o
-TCMODULES += q_qfq.o
-TCMODULES += m_gact.o
-TCMODULES += m_mirred.o
-TCMODULES += m_mpls.o
-TCMODULES += m_nat.o
-TCMODULES += m_pedit.o
-TCMODULES += m_ife.o
-TCMODULES += m_skbedit.o
-TCMODULES += m_skbmod.o
-TCMODULES += m_csum.o
-TCMODULES += m_simple.o
-TCMODULES += m_vlan.o
-TCMODULES += m_connmark.o
-TCMODULES += m_ctinfo.o
-TCMODULES += m_bpf.o
-TCMODULES += m_tunnel_key.o
-TCMODULES += m_sample.o
-TCMODULES += m_ct.o
-TCMODULES += p_ip.o
-TCMODULES += p_ip6.o
-TCMODULES += p_icmp.o
-TCMODULES += p_eth.o
-TCMODULES += p_tcp.o
-TCMODULES += p_udp.o
-TCMODULES += em_nbyte.o
-TCMODULES += em_cmp.o
-TCMODULES += em_u32.o
-TCMODULES += em_canid.o
-TCMODULES += em_meta.o
-TCMODULES += q_mqprio.o
-TCMODULES += q_codel.o
-TCMODULES += q_fq_codel.o
-TCMODULES += q_fq.o
-TCMODULES += q_pie.o
-TCMODULES += q_cake.o
-TCMODULES += q_hhf.o
-TCMODULES += q_clsact.o
-TCMODULES += e_bpf.o
-TCMODULES += f_matchall.o
-TCMODULES += q_cbs.o
-TCMODULES += q_etf.o
-TCMODULES += q_taprio.o
-TCMODULES += q_plug.o
+TCMODULES_y :=
+TCMODULES_m :=
+TCMODULES_$(CONFIG_NET_SCH_FIFO) += q_fifo.o
+TCMODULES_$(CONFIG_NET_SCH_SFQ) += q_sfq.o
+TCMODULES_$(CONFIG_NET_SCH_RED) += q_red.o
+TCMODULES_y += q_prio.o
+TCMODULES_$(CONFIG_NET_SCH_SKBPRIO) += q_skbprio.o
+TCMODULES_$(CONFIG_NET_SCH_TBF) += q_tbf.o
+TCMODULES_$(CONFIG_NET_SCH_CBQ) += q_cbq.o
+TCMODULES_$(CONFIG_NET_SCH_RR) += q_rr.o
+TCMODULES_$(CONFIG_NET_SCH_MULTIQ) += q_multiq.o
+TCMODULES_$(CONFIG_NET_SCH_NETEM) += q_netem.o
+TCMODULES_$(CONFIG_NET_SCH_CHOKE) += q_choke.o
+TCMODULES_$(CONFIG_NET_SCH_SFB) += q_sfb.o
+TCMODULES_$(CONFIG_NET_CLS_RSVP) += f_rsvp.o
+TCMODULES_$(CONFIG_NET_CLS_U32) += f_u32.o
+TCMODULES_$(CONFIG_NET_CLS_ROUTE4) += f_route.o
+TCMODULES_$(CONFIG_NET_CLS_FW) += f_fw.o
+TCMODULES_$(CONFIG_NET_CLS_BASIC) += f_basic.o
+TCMODULES_$(CONFIG_NET_CLS_BPF) += f_bpf.o
+TCMODULES_$(CONFIG_NET_CLS_FLOW) += f_flow.o
+TCMODULES_$(CONFIG_NET_CLS_CGROUP) += f_cgroup.o
+TCMODULES_$(CONFIG_NET_CLS_FLOWER) += f_flower.o
+TCMODULES_$(CONFIG_NET_SCH_DSMARK) += q_dsmark.o
+TCMODULES_$(CONFIG_NET_SCH_GRED) += q_gred.o
+TCMODULES_$(CONFIG_NET_CLS_TCINDEX) += f_tcindex.o
+TCMODULES_$(CONFIG_NET_SCH_INGRESS) += q_ingress.o
+TCMODULES_$(CONFIG_NET_SCH_HFSC) += q_hfsc.o
+TCMODULES_$(CONFIG_NET_SCH_HTB) += q_htb.o
+TCMODULES_$(CONFIG_NET_SCH_DRR) += q_drr.o
+TCMODULES_$(CONFIG_NET_SCH_QFQ) += q_qfq.o
+TCMODULES_$(CONFIG_NET_ACT_GACT) += m_gact.o
+TCMODULES_$(CONFIG_NET_ACT_MIRRED) += m_mirred.o
+TCMODULES_$(CONFIG_NET_ACT_MPLS) += m_mpls.o
+TCMODULES_$(CONFIG_NET_ACT_NAT) += m_nat.o
+TCMODULES_$(CONFIG_NET_ACT_PEDIT) += m_pedit.o
+TCMODULES_$(CONFIG_NET_ACT_IFE) += m_ife.o
+TCMODULES_$(CONFIG_NET_ACT_SKBEDIT) += m_skbedit.o
+TCMODULES_$(CONFIG_NET_ACT_SKBMOD) += m_skbmod.o
+TCMODULES_$(CONFIG_NET_ACT_CSUM) += m_csum.o
+TCMODULES_$(CONFIG_NET_ACT_SIMP) += m_simple.o
+TCMODULES_$(CONFIG_NET_ACT_VLAN) += m_vlan.o
+TCMODULES_$(CONFIG_NET_ACT_CONNMARK) += m_connmark.o
+TCMODULES_$(CONFIG_NET_ACT_CTINFO) += m_ctinfo.o
+TCMODULES_$(CONFIG_NET_ACT_BPF) += m_bpf.o
+TCMODULES_$(CONFIG_NET_ACT_TUNNEL_KEY) += m_tunnel_key.o
+TCMODULES_$(CONFIG_NET_ACT_SAMPLE) += m_sample.o
+TCMODULES_$(CONFIG_NET_ACT_CT) += m_ct.o
+TCMODULES_$(CONFIG_NET_ACT_PEDIT) += p_ip.o
+TCMODULES_$(CONFIG_NET_ACT_PEDIT) += p_ip6.o
+TCMODULES_$(CONFIG_NET_ACT_PEDIT) += p_icmp.o
+TCMODULES_$(CONFIG_NET_ACT_PEDIT) += p_eth.o
+TCMODULES_$(CONFIG_NET_ACT_PEDIT) += p_tcp.o
+TCMODULES_$(CONFIG_NET_ACT_PEDIT) += p_udp.o
+TCMODULES_$(CONFIG_NET_EMATCH_NBYTE) += em_nbyte.o
+TCMODULES_$(CONFIG_NET_EMATCH_CMP) += em_cmp.o
+TCMODULES_$(CONFIG_NET_EMATCH_U32) += em_u32.o
+TCMODULES_$(CONFIG_NET_EMATCH_CANID) += em_canid.o
+TCMODULES_$(CONFIG_NET_EMATCH_META) += em_meta.o
+TCMODULES_$(CONFIG_NET_SCH_MQPRIO) += q_mqprio.o
+TCMODULES_$(CONFIG_NET_SCH_CODEL) += q_codel.o
+TCMODULES_$(CONFIG_NET_SCH_FQ_CODEL) += q_fq_codel.o
+TCMODULES_$(CONFIG_NET_SCH_FQ) += q_fq.o
+TCMODULES_$(CONFIG_NET_SCH_PIE) += q_pie.o
+TCMODULES_$(CONFIG_NET_SCH_CAKE) += q_cake.o
+TCMODULES_$(CONFIG_NET_SCH_HHF) += q_hhf.o
+TCMODULES_y += q_clsact.o
+TCMODULES_$(CONFIG_BPF) += e_bpf.o
+TCMODULES_$(CONFIG_NET_CLS_MATCHALL) += f_matchall.o
+TCMODULES_$(CONFIG_NET_SCH_CBS) += q_cbs.o
+TCMODULES_$(CONFIG_NET_SCH_ETF) += q_etf.o
+TCMODULES_$(CONFIG_NET_SCH_TAPRIO) += q_taprio.o
+TCMODULES_$(CONFIG_NET_SCH_PLUG) += q_plug.o
 
 TCSO :=
 ifeq ($(TC_CONFIG_ATM),y)
@@ -86,11 +93,11 @@
 endif
 
 ifneq ($(TC_CONFIG_NO_XT),y)
-  ifeq ($(TC_CONFIG_XT),y)
+  ifeq ($(TC_CONFIG_XT)$(findstring y,$(CONFIG_NET_EMATCH_IPT)$(CONFIG_NET_EMATCH_IPSET)),yy)
     TCSO += m_xt.so
-    TCMODULES += em_ipt.o
+    TCMODULES_$(CONFIG_NET_EMATCH_IPT) += em_ipt.o
     ifeq ($(TC_CONFIG_IPSET),y)
-      TCMODULES += em_ipset.o
+      TCMODULES_$(CONFIG_NET_EMATCH_IPSET) += em_ipset.o
     endif
   else
     ifeq ($(TC_CONFIG_XT_OLD),y)
@@ -100,13 +107,13 @@
         CFLAGS += -DTC_CONFIG_XT_H
         TCSO += m_xt_old.so
       else
-        TCMODULES += m_ipt.o
+        TCMODULES_$(CONFIG_NET_MATCH_IPSET) += m_ipt.o
       endif
     endif
   endif
 endif
 
-TCOBJ += $(TCMODULES)
+TCOBJ += $(TCMODULES_y) $(TCMODULES_m)
 LDLIBS += -L. -lm
 
 ifeq ($(SHARED_LIBS),y)
