--- memtester-4.5.1/Makefile.org	2024-01-10 15:03:05.618487170 +1000
+++ memtester-4.5.1/Makefile	2024-01-10 15:04:45.307441376 +1000
@@ -10,8 +10,6 @@
 # You don't need to edit these; change the contents of the conf-cc and conf-ld
 # files if you need to change the compile/link commands.  See the README for
 # more information.
-CC			= $(shell head -n 1 conf-cc)
-LD			= $(shell head -n 1 conf-ld)
 
 SOURCES		= memtester.c tests.c
 OBJECTS		= $(SOURCES:.c=.o)
@@ -33,8 +31,8 @@
 auto-ccld.sh: \
 conf-cc conf-ld warn-auto.sh
 	( cat warn-auto.sh; \
-	echo CC=\'`head -1 conf-cc`\'; \
-	echo LD=\'`head -1 conf-ld`\' \
+	echo CC=\'"$(CC) $(CFLAGS) -DPOSIX -D_POSIX_C_SOURCE=200809L -D_FILE_OFFSET_BITS=64 -DTEST_NARROW_WRITES -c"\' ; \
+	echo LD=\'"$(CC) $(CFLAGS) -s"\' \
 	) > auto-ccld.sh
 
 compile: \
