diff -ru aircrack-ng-1.2-rc4.orig/autocfg aircrack-ng-1.2-rc4/autocfg
--- aircrack-ng-1.2-rc4.orig/autocfg	2016-02-08 15:12:08.000000000 +1000
+++ aircrack-ng-1.2-rc4/autocfg	2017-09-21 11:35:02.408789666 +1000
@@ -116,7 +116,7 @@
 }
 
 case "$CC" in
-	mips-* | arm-*)
+	*mips-* | *arm-*)
 		IS_CROSS=1
 		IS_X86=0
 	;;
Only in aircrack-ng-1.2-rc4: autocfg.orig
diff -ru aircrack-ng-1.2-rc4.orig/common.mak aircrack-ng-1.2-rc4/common.mak
--- aircrack-ng-1.2-rc4.orig/common.mak	2016-01-01 05:42:10.000000000 +1000
+++ aircrack-ng-1.2-rc4/common.mak	2017-09-21 11:42:01.869309156 +1000
@@ -207,7 +207,7 @@
 
 CFLAGS          += $(OPTFLAGS) $(REVFLAGS) $(COMMON_CFLAGS)
 
-prefix          = /usr/local
+prefix          = /
 bindir          = $(prefix)/bin
 sbindir         = $(prefix)/sbin
 mandir          = $(prefix)/share/man/man1
@@ -231,16 +231,6 @@
 	GCC_OVER49	= $(shell expr 4.9 \<= `$(CC) -dumpversion | awk -F. '{ print $1$2 }'`)
 endif
 
-ifeq ($(GCC_OVER49), 0)
-	ifeq ($(GCC_OVER41), 1)
-		COMMON_CFLAGS += -fstack-protector
-	endif
-endif
-
-ifeq ($(GCC_OVER49), 1)
-	COMMON_CFLAGS += -fstack-protector-strong
-endif
-
 ifeq ($(GCC_OVER45), 1)
 	CFLAGS		+= -Wno-unused-but-set-variable -Wno-array-bounds
 endif
diff -ru aircrack-ng-1.2-rc4.orig/src/aircrack-ng.h aircrack-ng-1.2-rc4/src/aircrack-ng.h
--- aircrack-ng-1.2-rc4.orig/src/aircrack-ng.h	2016-02-08 13:23:16.000000000 +1000
+++ aircrack-ng-1.2-rc4/src/aircrack-ng.h	2017-09-21 11:35:53.068723317 +1000
@@ -143,14 +143,14 @@
 	A_neg						 /* helps reject false positives */
 };
 
-struct dictfiles {
+static struct dictfiles {
 	off_t	dictsize;			/* Total file size */
 	off_t	dictpos;			/* Current position of dictionary */
 	off_t	wordcount;			/* Total amount of words in dict file */
 	int	loaded;				/* Have finished processing? */
 } dicts;
 
-struct options
+static struct options
 {
 	int amode;					 /* attack mode          */
 	int essid_set;				 /* essid set flag       */
@@ -230,7 +230,7 @@
 typedef struct { int idx, val; }
 vote;
 
-struct WEP_data
+static struct WEP_data
 {
 	unsigned char key[64];				 /* the current chosen WEP key   */
 	unsigned char *ivbuf;				 /* buffer holding all the IVs   */
diff -ru aircrack-ng-1.2-rc4.orig/src/linecount.cpp aircrack-ng-1.2-rc4/src/linecount.cpp
--- aircrack-ng-1.2-rc4.orig/src/linecount.cpp	2016-02-15 09:34:57.000000000 +1000
+++ aircrack-ng-1.2-rc4/src/linecount.cpp	2017-09-21 11:35:03.488787773 +1000
@@ -54,6 +54,7 @@
  *  find a better solution performance wise.
  */
 
+#include <unistd.h>
 #include <iostream>
 #include <fstream>
 #include <vector>
