diff -rbc libpciaccess-0.13.4/src/Makefile.in libpciaccess-0.13.4-patched/src/Makefile.in
*** libpciaccess-0.13.4/src/Makefile.in	2015-05-01 14:44:58.000000000 +1000
--- libpciaccess-0.13.4-patched/src/Makefile.in	2015-10-26 16:35:39.785058189 +1000
***************
*** 417,423 ****
  	$(OS_SUPPORT)
  
  libpciaccess_la_LIBADD = $(PCIACCESS_LIBS)
! libpciaccess_la_LDFLAGS = -version-number 0:11:1 -no-undefined
  all: all-am
  
  .SUFFIXES:
--- 417,423 ----
  	$(OS_SUPPORT)
  
  libpciaccess_la_LIBADD = $(PCIACCESS_LIBS)
! libpciaccess_la_LDFLAGS =
  all: all-am
  
  .SUFFIXES:
