diff --git a/shellinabox/root_page.html b/shellinabox/root_page.html
index a04ea21..3b9a327 100644
--- a/shellinabox/root_page.html
+++ b/shellinabox/root_page.html
@@ -117,7 +117,7 @@
        correctly deal with the enclosing frameset (if any), if we do not
        do this
    -->
-  <body onload="setTimeout('window.shellinabox = new ShellInABox()', 100)"
+  <body onload="setTimeout('window.shellinabox = new ShellInABox(location.href)', 100)"
         scroll="no"><noscript>JavaScript
     must be enabled for ShellInABox</noscript></body>
 </html>
diff --git a/shellinabox/shell_in_a_box.jspp b/shellinabox/shell_in_a_box.jspp
index 4e449cc..1b152bc 100644
--- a/shellinabox/shell_in_a_box.jspp
+++ b/shellinabox/shell_in_a_box.jspp
@@ -170,7 +170,7 @@ ShellInABox.prototype.sendRequest = function(request) {
   if (request == undefined) {
     request                  = new XMLHttpRequest();
   }
-  request.open('POST', this.url + '?', true);
+  request.open('POST', this.url, true);
   request.timeout = 30000; // Don't leave POST pending forever: force 30s timeout to prevent HTTP Proxy thread hijack
   request.setRequestHeader('Cache-Control', 'no-cache');
   request.setRequestHeader('Content-Type',
@@ -239,7 +239,7 @@ ShellInABox.prototype.sendKeys = function(keys) {
     keys                       = this.pendingKeys + keys;
     this.pendingKeys           = '';
     var request                = new XMLHttpRequest();
-    request.open('POST', this.url + '?', true);
+    request.open('POST', this.url, true);
     request.setRequestHeader('Cache-Control', 'no-cache');
     request.setRequestHeader('Content-Type',
                            'application/x-www-form-urlencoded; charset=utf-8');
