--- a/libhttp/ssl.c.orig	2022-09-13 10:44:19.189959195 +1000
+++ b/libhttp/ssl.c	2022-09-13 10:47:44.718309342 +1000
@@ -1026,50 +1026,8 @@
 void sslFreeHndl(SSL **sslHndl) {
 #if defined(HAVE_OPENSSL)
   if (*sslHndl) {
-    // OpenSSL does not always correctly perform reference counting for stacked
-    // BIOs. This is particularly a problem if an SSL connection has two
-    // different BIOs for the read and the write end, with one being a stacked
-    // derivative of the other. Unfortunately, this is exactly the scenario
-    // that we set up with call to "BIO_push(readBIO, writeBIO)" in function
-    // "sslPromoteToSSL()".
-    // As a work-around, we un-stack the BIOs prior to freeing the SSL
-    // connection.
     debug("[ssl] Freeing SSL handle.");
     ERR_clear_error();
-    BIO *writeBIO, *readBIO;
-    check(writeBIO    = SSL_get_wbio(*sslHndl));
-    check(readBIO     = SSL_get_rbio(*sslHndl));
-    if (writeBIO != readBIO) {
-      if (sslGetNextBIO(readBIO) == writeBIO) {
-        // OK, that's exactly the bug we are looking for. We know that
-        // writeBIO needs to be removed from readBIO chain.
-        debug("[ssl] Removing stacked write BIO!");
-        check(BIO_pop(readBIO) == writeBIO);
-        check(!sslGetNextBIO(readBIO));
-      } else if (sslGetNextBIO(readBIO) == sslGetNextBIO(writeBIO)) {
-        // Things get even more confused, if the SSL handshake is aborted
-        // prematurely.
-        // OpenSSL appears to internally stack a BIO onto the read end that
-        // does not get removed afterwards. We end up with the original
-        // socket BIO having two different BIOs prepended to it (one for
-        // reading and one for writing). In this case, not only is the
-        // reference count wrong, but the chain of next_bio/prev_bio pairs
-        // is corrupted, too.
-        warn("[ssl] Removing stacked socket BIO!");
-        BIO *sockBIO;
-        check(sockBIO = BIO_pop(readBIO));
-        check(sockBIO == BIO_pop(writeBIO));
-        check(!sslGetNextBIO(readBIO));
-        check(!sslGetNextBIO(writeBIO));
-        check(!sslGetNextBIO(sockBIO));
-        BIO_free_all(sockBIO);
-      } else {
-        // We do not know, how to fix this situation. Something must have
-        // changed in the OpenSSL internals. Either, this is a new bug, or
-        // somebody fixed the code in a way that we did not anticipate.
-        fatal("[ssl] Unexpected corruption of OpenSSL data structures");
-      }
-    }
     SSL_free(*sslHndl);
     dcheck(!ERR_peek_error());
   }
