diff -Naur libnetfilter_queue-1.0.2.org/src/internal.h libnetfilter_queue-1.0.2/src/internal.h
--- libnetfilter_queue-1.0.2.org/src/internal.h	2012-08-06 22:50:10.000000000 +1000
+++ libnetfilter_queue-1.0.2/src/internal.h	2014-02-21 13:12:08.000000000 +1000
@@ -6,7 +6,7 @@
 #	define __visible	__attribute__((visibility("default")))
 #	define EXPORT_SYMBOL(x)	typeof(x) (x) __visible
 #else
-#	define EXPORT_SYMBOL
+#	define EXPORT_SYMBOL(x)
 #endif
 
 struct iphdr;
