diff -u ethtool-3.13.orig/ethtool.c ethtool-3.13/ethtool.c
--- ethtool-3.13.orig/ethtool.c	2014-01-28 09:54:09.000000000 +1000
+++ ethtool-3.13/ethtool.c	2014-02-05 18:05:41.131728378 +1000
@@ -872,6 +872,7 @@
 	return buf;
 }
 
+#if 0
 static const struct {
 	const char *name;
 	int (*func)(struct ethtool_drvinfo *info, struct ethtool_regs *regs);
@@ -904,6 +905,7 @@
 	{ "st_gmac", st_gmac_dump_regs },
 	{ "et131x", et131x_dump_regs },
 };
+#endif
 
 void dump_hex(FILE *file, const u8 *data, int len, int offset)
 {
@@ -923,7 +925,9 @@
 		     const char *gregs_dump_file,
 		     struct ethtool_drvinfo *info, struct ethtool_regs *regs)
 {
+#if 0
 	int i;
+#endif
 
 	if (gregs_dump_raw) {
 		fwrite(regs->data, regs->len, 1, stdout);
@@ -946,6 +950,7 @@
 		fclose(f);
 	}
 
+#if 0
 	if (!gregs_dump_hex)
 		for (i = 0; i < ARRAY_SIZE(driver_list); i++)
 			if (!strncmp(driver_list[i].name, info->driver,
@@ -958,6 +963,7 @@
 				 */
 				break;
 			}
+#endif
 
 	dump_hex(stdout, regs->data, regs->len, 0);
 
@@ -972,11 +978,13 @@
 		return 0;
 	}
 
+#if 0
 	if (!strncmp("natsemi", info->driver, ETHTOOL_BUSINFO_LEN)) {
 		return natsemi_dump_eeprom(info, ee);
 	} else if (!strncmp("tg3", info->driver, ETHTOOL_BUSINFO_LEN)) {
 		return tg3_dump_eeprom(info, ee);
 	}
+#endif
 
 	dump_hex(stdout, ee->data, ee->len, ee->offset);
 
@@ -3628,6 +3636,7 @@
 	if (geeprom_dump_raw) {
 		fwrite(eeprom->data, 1, eeprom->len, stdout);
 	} else {
+#if 0
 		if (eeprom->offset != 0  ||
 		    (eeprom->len != modinfo.eeprom_len)) {
 			geeprom_dump_hex = 1;
@@ -3646,6 +3655,7 @@
 			}
 		}
 		if (geeprom_dump_hex)
+#endif
 			dump_hex(stdout, eeprom->data,
 				 eeprom->len, eeprom->offset);
 	}
diff -u ethtool-3.13.orig/Makefile.am ethtool-3.13/Makefile.am
--- ethtool-3.13.orig/Makefile.am	2014-01-27 11:27:14.000000000 +1000
+++ ethtool-3.13/Makefile.am	2014-02-05 17:59:23.995739543 +1000
@@ -6,11 +6,7 @@
 
 sbin_PROGRAMS = ethtool
 ethtool_SOURCES = ethtool.c ethtool-copy.h internal.h net_tstamp-copy.h \
-		  amd8111e.c de2104x.c e100.c e1000.c et131x.c igb.c	\
-		  fec_8xx.c ibm_emac.c ixgb.c ixgbe.c natsemi.c	\
-		  pcnet32.c realtek.c tg3.c marvell.c vioc.c	\
-		  smsc911x.c at76c50x-usb.c sfc.c stmmac.c	\
-		  rxclass.c sfpid.c sfpdiag.c ixgbevf.c
+		  rxclass.c
 
 TESTS = test-cmdline test-features
 check_PROGRAMS = test-cmdline test-features
diff -u ethtool-3.13.orig/Makefile.in ethtool-3.13/Makefile.in
--- ethtool-3.13.orig/Makefile.in	2014-01-28 10:01:34.000000000 +1000
+++ ethtool-3.13/Makefile.in	2014-02-05 18:03:08.595732894 +1000
@@ -98,32 +98,12 @@
 CONFIG_CLEAN_VPATH_FILES =
 am__installdirs = "$(DESTDIR)$(sbindir)" "$(DESTDIR)$(man8dir)"
 PROGRAMS = $(sbin_PROGRAMS)
-am_ethtool_OBJECTS = ethtool.$(OBJEXT) amd8111e.$(OBJEXT) \
-	de2104x.$(OBJEXT) e100.$(OBJEXT) e1000.$(OBJEXT) \
-	et131x.$(OBJEXT) igb.$(OBJEXT) fec_8xx.$(OBJEXT) \
-	ibm_emac.$(OBJEXT) ixgb.$(OBJEXT) ixgbe.$(OBJEXT) \
-	natsemi.$(OBJEXT) pcnet32.$(OBJEXT) realtek.$(OBJEXT) \
-	tg3.$(OBJEXT) marvell.$(OBJEXT) vioc.$(OBJEXT) \
-	smsc911x.$(OBJEXT) at76c50x-usb.$(OBJEXT) sfc.$(OBJEXT) \
-	stmmac.$(OBJEXT) rxclass.$(OBJEXT) sfpid.$(OBJEXT) \
-	sfpdiag.$(OBJEXT) ixgbevf.$(OBJEXT)
+am_ethtool_OBJECTS = ethtool.$(OBJEXT) rxclass.$(OBJEXT)
 ethtool_OBJECTS = $(am_ethtool_OBJECTS)
 ethtool_LDADD = $(LDADD)
 ethtool_DEPENDENCIES =
 am__objects_1 = test_cmdline-ethtool.$(OBJEXT) \
-	test_cmdline-amd8111e.$(OBJEXT) test_cmdline-de2104x.$(OBJEXT) \
-	test_cmdline-e100.$(OBJEXT) test_cmdline-e1000.$(OBJEXT) \
-	test_cmdline-et131x.$(OBJEXT) test_cmdline-igb.$(OBJEXT) \
-	test_cmdline-fec_8xx.$(OBJEXT) test_cmdline-ibm_emac.$(OBJEXT) \
-	test_cmdline-ixgb.$(OBJEXT) test_cmdline-ixgbe.$(OBJEXT) \
-	test_cmdline-natsemi.$(OBJEXT) test_cmdline-pcnet32.$(OBJEXT) \
-	test_cmdline-realtek.$(OBJEXT) test_cmdline-tg3.$(OBJEXT) \
-	test_cmdline-marvell.$(OBJEXT) test_cmdline-vioc.$(OBJEXT) \
-	test_cmdline-smsc911x.$(OBJEXT) \
-	test_cmdline-at76c50x-usb.$(OBJEXT) test_cmdline-sfc.$(OBJEXT) \
-	test_cmdline-stmmac.$(OBJEXT) test_cmdline-rxclass.$(OBJEXT) \
-	test_cmdline-sfpid.$(OBJEXT) test_cmdline-sfpdiag.$(OBJEXT) \
-	test_cmdline-ixgbevf.$(OBJEXT)
+	test_cmdline-rxclass.$(OBJEXT)
 am_test_cmdline_OBJECTS = test_cmdline-test-cmdline.$(OBJEXT) \
 	test_cmdline-test-common.$(OBJEXT) $(am__objects_1)
 test_cmdline_OBJECTS = $(am_test_cmdline_OBJECTS)
@@ -132,21 +112,7 @@
 test_cmdline_LINK = $(CCLD) $(test_cmdline_CFLAGS) $(CFLAGS) \
 	$(AM_LDFLAGS) $(LDFLAGS) -o $@
 am__objects_2 = test_features-ethtool.$(OBJEXT) \
-	test_features-amd8111e.$(OBJEXT) \
-	test_features-de2104x.$(OBJEXT) test_features-e100.$(OBJEXT) \
-	test_features-e1000.$(OBJEXT) test_features-et131x.$(OBJEXT) \
-	test_features-igb.$(OBJEXT) test_features-fec_8xx.$(OBJEXT) \
-	test_features-ibm_emac.$(OBJEXT) test_features-ixgb.$(OBJEXT) \
-	test_features-ixgbe.$(OBJEXT) test_features-natsemi.$(OBJEXT) \
-	test_features-pcnet32.$(OBJEXT) \
-	test_features-realtek.$(OBJEXT) test_features-tg3.$(OBJEXT) \
-	test_features-marvell.$(OBJEXT) test_features-vioc.$(OBJEXT) \
-	test_features-smsc911x.$(OBJEXT) \
-	test_features-at76c50x-usb.$(OBJEXT) \
-	test_features-sfc.$(OBJEXT) test_features-stmmac.$(OBJEXT) \
-	test_features-rxclass.$(OBJEXT) test_features-sfpid.$(OBJEXT) \
-	test_features-sfpdiag.$(OBJEXT) \
-	test_features-ixgbevf.$(OBJEXT)
+	test_features-rxclass.$(OBJEXT)
 am_test_features_OBJECTS = test_features-test-features.$(OBJEXT) \
 	test_features-test-common.$(OBJEXT) $(am__objects_2)
 test_features_OBJECTS = $(am_test_features_OBJECTS)
@@ -533,11 +499,7 @@
 man_MANS = ethtool.8
 EXTRA_DIST = LICENSE ethtool.8 ethtool.spec.in aclocal.m4 ChangeLog autogen.sh
 ethtool_SOURCES = ethtool.c ethtool-copy.h internal.h net_tstamp-copy.h \
-		  amd8111e.c de2104x.c e100.c e1000.c et131x.c igb.c	\
-		  fec_8xx.c ibm_emac.c ixgb.c ixgbe.c natsemi.c	\
-		  pcnet32.c realtek.c tg3.c marvell.c vioc.c	\
-		  smsc911x.c at76c50x-usb.c sfc.c stmmac.c	\
-		  rxclass.c sfpid.c sfpdiag.c ixgbevf.c
+		  rxclass.c
 
 test_cmdline_SOURCES = test-cmdline.c test-common.c $(ethtool_SOURCES) 
 test_cmdline_CFLAGS = -DTEST_ETHTOOL
