diff -ur grub-2.02.orig/grub-core/commands/probe.c grub-2.02/grub-core/commands/probe.c
--- grub-2.02.orig/grub-core/commands/probe.c	2015-05-22 01:50:29.000000000 +1000
+++ grub-2.02/grub-core/commands/probe.c	2019-03-15 14:43:24.714389763 +1000
@@ -45,6 +45,7 @@
     {"fs",		'f', 0, N_("Determine filesystem type."), 0, 0},
     {"fs-uuid",		'u', 0, N_("Determine filesystem UUID."), 0, 0},
     {"label",		'l', 0, N_("Determine filesystem label."), 0, 0},
+    {"part-uuid",	'U', 0, N_("Determine partition UUID (GPT)."), 0, 0},
     {0, 0, 0, 0, 0, 0}
   };
 
@@ -94,6 +95,18 @@
       if (state[0].set)
 	grub_env_set (state[0].arg, val);
       else
+	grub_printf ("%s", val);
+      grub_device_close (dev);
+      return GRUB_ERR_NONE;
+    }
+  if (state[6].set)
+    {
+      const char *val = "none";
+      if (dev->disk && dev->disk->partition)
+	val = dev->disk->partition->uuid;
+      if (state[0].set)
+	grub_env_set (state[0].arg, val);
+      else
 	grub_printf ("%s", val);
       grub_device_close (dev);
       return GRUB_ERR_NONE;
diff -ur grub-2.02.orig/grub-core/partmap/gpt.c grub-2.02/grub-core/partmap/gpt.c
--- grub-2.02.orig/grub-core/partmap/gpt.c	2015-09-03 06:10:53.000000000 +1000
+++ grub-2.02/grub-core/partmap/gpt.c	2019-03-15 15:09:57.322424322 +1000
@@ -109,6 +109,17 @@
 	  part.partmap = &grub_gpt_partition_map;
 	  part.parent = disk->partition;
 
+	  /* copy uuid across */
+          part.uuid = grub_xasprintf ("%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x",
+			entry.guid[3],  entry.guid[2],
+			entry.guid[1],  entry.guid[0],
+			entry.guid[5],  entry.guid[4],
+			entry.guid[7],  entry.guid[6],
+			entry.guid[8],  entry.guid[9],
+			entry.guid[10], entry.guid[11],
+			entry.guid[12], entry.guid[13],
+			entry.guid[14], entry.guid[15] );
+
 	  grub_dprintf ("gpt", "GPT entry %d: start=%lld, length=%lld\n", i,
 			(unsigned long long) part.start,
 			(unsigned long long) part.len);
diff -ur grub-2.02.orig/include/grub/partition.h grub-2.02/include/grub/partition.h
--- grub-2.02.orig/include/grub/partition.h	2015-05-22 01:50:29.000000000 +1000
+++ grub-2.02/include/grub/partition.h	2019-03-15 14:51:15.466399978 +1000
@@ -87,6 +87,9 @@
   /* The type of partition whne it's on MSDOS.
      Used for embedding detection.  */
   grub_uint8_t msdostype;
+
+  /* partition UUID for linux/GPT/EFI/etc */
+  char *uuid;
 };
 
 grub_partition_t EXPORT_FUNC(grub_partition_probe) (struct grub_disk *disk,
