From: David McCullough <david.mccullough@accelecon.com>
Date: Tue, 6 Aug 2013 14:36:52 +1000
Subject: build: allow old autoconf version

---
 configure.ac | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/configure.ac b/configure.ac
index 7259ed1..4c4a61a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1,5 +1,5 @@
 dnl Process this file with autoconf to produce a configure script.
-AC_PREREQ([2.68])
+dnl AC_PREREQ([2.68])
 
 dnl The QMI version number
 m4_define([qmi_major_version], [1])
@@ -52,7 +52,7 @@ m4_pattern_forbid([^AX_(COMPILER_FLAGS|COMPILER_FLAGS_(CFLAGS|GIR|LDFLAGS))\b],
   [Unexpanded AX_ macro found. Please install GNU autoconf-archive])
 
 dnl Detect git or release builds
-AX_IS_RELEASE([git-directory])
+dnl AX_IS_RELEASE([git-directory])
 
 dnl Function type cast disabled: used throughout the code especially to
 dnl cast GAsyncReadyCallbacks with the real object type instead of GObject
@@ -62,8 +62,8 @@ dnl All message protocol structs are packed, never complain about it
 DISABLED_WARNINGS="${DISABLED_WARNINGS} -Wno-packed"
 
 dnl Setup compiler checks
-AX_COMPILER_FLAGS()
-AX_COMPILER_FLAGS_CFLAGS(,,,[${DISABLED_WARNINGS}])
+dnl AX_COMPILER_FLAGS()
+dnl AX_COMPILER_FLAGS_CFLAGS(,,,[${DISABLED_WARNINGS}])
 
 dnl Specify gnu89 mode
 if test "$GCC" = "yes"; then
