diff --git a/src/qmi-firmware-update/qfu-updater.c b/src/qmi-firmware-update/qfu-updater.c
index 7fe1112..58c59af 100644
--- a/src/qmi-firmware-update/qfu-updater.c
+++ b/src/qmi-firmware-update/qfu-updater.c
@@ -1042,7 +1042,6 @@ validate_firmware_config_carrier (QfuUpdater  *self,
         g_warning ("[qfu-updater] user provided carrier doesn't match the one in the specified images: "
                    "'%s' != '%s' (IGNORED with --ignore-version-errors)",
                    self->priv->carrier, ctx->carrier);
-        return FALSE;
     }
 
     /* No firmware version? */
