diff --git a/src/qmi-firmware-update/qfu-udev-helpers.c b/src/qmi-firmware-update/qfu-udev-helpers.c
index c789703..024add0 100644
--- a/src/qmi-firmware-update/qfu-udev-helpers.c
+++ b/src/qmi-firmware-update/qfu-udev-helpers.c
@@ -380,8 +380,9 @@ device_matches_sysfs_and_type (GUdevDevice             *device,
     if (!device_sysfs_path)
         goto out;
 
-    if (g_strcmp0 (device_sysfs_path, sysfs_path) != 0)
-        goto out;
+    // Don't match the device path - some modems (e.g. MC7430) have different device paths when in QDL mode
+    // if (g_strcmp0 (device_sysfs_path, sysfs_path) != 0)
+    //     goto out;
 
     if (!udev_helper_get_udev_interface_details (device,
                                                  &device_driver,
