From cc8b1b1449c6f61f0232d0bf9fc1c1c0f5c66eef Mon Sep 17 00:00:00 2001
From: "Sreeves, Francis" <francis.sreeves@digi.com>
Date: Thu, 12 Jan 2023 10:13:03 +0000
Subject: [PATCH] sys_tree: $SYS/broker/clients/disconnected should never be
 negative

---
 src/sys_tree.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/sys_tree.c b/src/sys_tree.c
index 49ab7c9..b1eee6b 100644
--- a/src/sys_tree.c
+++ b/src/sys_tree.c
@@ -87,7 +87,11 @@ static void sys_tree__update_clients(char *buf)
 	}
 
 	if(disconnected_count != count_total-count_by_sock){
-		disconnected_count = count_total-count_by_sock;
+		/* If a client has connected but not sent a CONNECT at this point,
+		 * then it is possible that count_by_sock will be bigger than
+		 * count_total, causing a negative number. This situation should
+		 * not last for long, so just cap at zero and ignore. */
+		disconnected_count = count_total > count_by_sock ? count_total-count_by_sock : 0;
 		len = (uint32_t)snprintf(buf, BUFLEN, "%d", disconnected_count);
 		db__messages_easy_queue(NULL, "$SYS/broker/clients/inactive", SYS_TREE_QOS, len, buf, 1, SYS_TREE_EXPIRY, NULL);
 		db__messages_easy_queue(NULL, "$SYS/broker/clients/disconnected", SYS_TREE_QOS, len, buf, 1, SYS_TREE_EXPIRY, NULL);
-- 
2.34.1

