diff -Naur udev-182.org/src/libudev.h udev-182/src/libudev.h
--- udev-182.org/src/libudev.h	2012-01-29 10:15:46.960827009 +1000
+++ udev-182/src/libudev.h	2019-06-13 16:40:45.214028212 +1000
@@ -14,6 +14,7 @@
 
 #include <stdarg.h>
 #include <sys/types.h>
+#include <sys/sysmacros.h>
 #include <sys/stat.h>
 
 #ifdef __cplusplus
diff -Naur udev-182.org/src/udev.h udev-182/src/udev.h
--- udev-182.org/src/udev.h	2012-01-29 14:32:00.763033394 +1000
+++ udev-182/src/udev.h	2019-06-13 16:40:27.326738339 +1000
@@ -20,6 +20,7 @@
 #define _UDEV_H_
 
 #include <sys/types.h>
+#include <sys/sysmacros.h>
 #include <sys/param.h>
 #include <signal.h>
 
