From: Thomas Petazzoni <thomas.petazzoni@free-electrons.com>
Date: Tue, 7 Mar 2017 23:29:05 +0100
Subject: [PATCH 26/31] Add an option to disable the readline module

Signed-off-by: Thomas Petazzoni <thomas.petazzoni@free-electrons.com>
---
 configure.ac | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/configure.ac b/configure.ac
index bc2675b..d3875df 100644
--- a/configure.ac
+++ b/configure.ac
@@ -3178,6 +3178,14 @@ AC_ARG_ENABLE(openssl,
 	  fi])
 AC_MSG_RESULT($DISABLED_EXTENSIONS)
 
+AC_MSG_CHECKING(for --disable-readline)
+AC_ARG_ENABLE(readline,
+	AS_HELP_STRING([--disable-readline], [disable readline]),
+	[ if test "$enableval" = "no"; then
+	     DISABLED_EXTENSIONS="${DISABLED_EXTENSIONS} readline"
+	  fi])
+AC_MSG_RESULT($DISABLED_EXTENSIONS)
+
 AC_SUBST(TK)
 AC_MSG_CHECKING(for --disable-tk)
 AC_ARG_ENABLE(tk,
