From: Nicolas Cavallari <nicolas.cavallari@green-communications.fr>
Date: Wed, 22 Feb 2017 17:55:59 -0800
Subject: [PATCH 25/31] Add an option to disable openssl support.

Signed-off-by: Nicolas Cavallari <nicolas.cavallari@green-communications.fr>
---
 configure.ac | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/configure.ac b/configure.ac
index e68dc79..bc2675b 100644
--- a/configure.ac
+++ b/configure.ac
@@ -3170,6 +3170,14 @@ AC_ARG_ENABLE(unicodedata,
   	  fi])
 AC_MSG_RESULT($DISABLED_EXTENSIONS)
 
+AC_MSG_CHECKING(for --disable-openssl)
+AC_ARG_ENABLE(openssl,
+	AS_HELP_STRING([--disable-openssl], [disable openssl support]),
+	[ if test "$enableval" = "no"; then
+	     DISABLED_EXTENSIONS="${DISABLED_EXTENSIONS} ssl _ssl _hashlib"
+	  fi])
+AC_MSG_RESULT($DISABLED_EXTENSIONS)
+
 AC_SUBST(TK)
 AC_MSG_CHECKING(for --disable-tk)
 AC_ARG_ENABLE(tk,
