diff --git a/mDNSEmbedLib/Makefile b/mDNSEmbedLib/Makefile
new file mode 100644
index 0000000..237e592
--- /dev/null
+++ b/mDNSEmbedLib/Makefile
@@ -0,0 +1,50 @@
+INCDIRS = ../mDNSCore ../mDNSPosix ../mDNSShared
+SRCDIRS = ../mDNSCore ../mDNSPosix ../mDNSShared
+INCLUDES=  $(wildcard *.h) $(foreach d, $(INCDIRS), $(wildcard $d/*.h))
+SOURCES = $(wildcard *.c*)
+CFLAGS += -D_GNU_SOURCE -DNOT_HAVE_SA_LEN -D_POSIX -DUSES_NETLINK -DHAVE_LINUX -DTARGET_OS_LINUX -DMDNS_DEBUGMSGS=0 -W -Wall -Werror -fno-strict-aliasing -fdata-sections -ffunction-sections
+CFLAGS+=$(addprefix -I,$(INCDIRS)) -fPIC
+TARGETLIB = libmdnsembed.so
+
+vpath %.c $(SRCDIRS)
+vpath %.S $(SRCDIRS)
+vpath %.s $(SRCDIRS)
+vpath %.h $(INCDIRS)
+
+SOURCES += \
+	mDNSPosix.c \
+	mDNSUNP.c \
+	mDNS.c \
+	uDNS.c \
+	DNSCommon.c \
+	DNSDigest.c \
+	anonymous.c \
+	CryptoAlg.c \
+	mDNSDebug.c \
+	PlatformCommon.c \
+	GenLinkedList.c \
+	$()
+
+OBJECTS  = $(patsubst %.c,%.o, $(notdir $(SOURCES)))
+
+default: $(TARGETLIB)
+
+%.o: %.c $(notdir $(INCLUDES)) $(OBJECTS)
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+.PHONY: clean default
+
+$(TARGETLIB): $(OBJECTS)
+	$(CC) -shared -o $@ $^ $(LDFLAGS)
+
+install: $(TARGETLIB)
+	install -d -m 0755 $(DESTDIR)/lib
+	install -d -m 0755 $(DESTDIR)/include
+	install -m 0755 $(TARGETLIB) $(DESTDIR)/lib/$(TARGETLIB)
+	install -m 0755 ../mDNSCore/mDNSEmbeddedAPI.h $(DESTDIR)/include/mDNSEmbeddedAPI.h
+	install -m 0755 ../mDNSCore/mDNSDebug.h $(DESTDIR)/include/mDNSDebug.h
+	install -m 0755 ../mDNSPosix/mDNSPosix.h $(DESTDIR)/include/mDNSPosix.h
+
+clean:
+	rm -rf $(TARGETLIB) *.o
+
