diff -ur openssl-1.0.2p.orig/Configure openssl-1.0.2p/Configure
--- openssl-1.0.2p.orig/Configure	2015-01-23 00:58:32.000000000 +1000
+++ openssl-1.0.2p/Configure	2015-03-04 13:48:04.264916581 +1000
@@ -680,8 +680,8 @@
 "tandem-c89","c89:-Ww -D__TANDEM -D_XOPEN_SOURCE -D_XOPEN_SOURCE_EXTENDED=1 -D_TANDEM_SOURCE -DB_ENDIAN::(unknown):::THIRTY_TWO_BIT:::",
 
 # uClinux
-"uClinux-dist","$ENV{'CC'}:\$(CFLAGS)::-D_REENTRANT::\$(LDFLAGS) \$(LDLIBS):BN_LLONG:${no_asm}:$ENV{'LIBSSL_dlfcn'}:linux-shared:-fPIC:-shared:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR):$ENV{'RANLIB'}::",
-"uClinux-dist64","$ENV{'CC'}:\$(CFLAGS)::-D_REENTRANT::\$(LDFLAGS) \$(LDLIBS):SIXTY_FOUR_BIT_LONG:${no_asm}:$ENV{'LIBSSL_dlfcn'}:linux-shared:-fPIC:-shared:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR):$ENV{'RANLIB'}::",
+"uClinux-dist","$ENV{'CC'}:\$(CFLAGS)::-D_REENTRANT::\$(LDFLAGS) \$(LDLIBS):BN_LLONG:".(eval "\$$ENV{'UCDIST_ASM'}").":$ENV{'LIBSSL_dlfcn'}:linux-shared:-fPIC:-shared:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR):$ENV{'RANLIB'}::",
+"uClinux-dist64","$ENV{'CC'}:\$(CFLAGS)::-D_REENTRANT::\$(LDFLAGS) \$(LDLIBS):SIXTY_FOUR_BIT_LONG:".(eval "\$$ENV{'UCDIST_ASM'}").":\$(UCDIST_TYP):$ENV{'LIBSSL_dlfcn'}:linux-shared:-fPIC:-shared:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR):$ENV{'RANLIB'}::",
 
 );
 
diff -ur openssl-1.0.2p.orig/crypto/arm_arch.h openssl-1.0.2p/crypto/arm_arch.h
--- openssl-1.0.2p.orig/crypto/arm_arch.h	2015-01-23 00:58:32.000000000 +1000
+++ openssl-1.0.2p/crypto/arm_arch.h	2015-03-04 13:51:06.976919536 +1000
@@ -42,6 +42,8 @@
 #    define __ARM_ARCH__ 5
 #   elif defined(__ARM_ARCH_4__) || defined(__ARM_ARCH_4T__)
 #    define __ARM_ARCH__ 4
+#   elif defined(__ARM_ARCH_3__) || defined(__ARM_ARCH_3T__)
+#    define __ARM_ARCH__ 3
 #   else
 #    error "unsupported ARM architecture"
 #   endif
--- openssl-1.0.2p.orig/crypto/evp/c_all.c	2018-04-13 14:06:39.955785330 +1000
+++ openssl-1.0.2p/crypto/evp/c_all.c	2018-04-13 14:06:59.035785744 +1000
@@ -82,4 +82,7 @@
     OPENSSL_cpuid_setup();
     OpenSSL_add_all_ciphers();
     OpenSSL_add_all_digests();
+#if defined(OPENSSL_NO_HW) && (defined(__OpenBSD__) || defined(__FreeBSD__) || defined(HAVE_CRYPTODEV))
+    ENGINE_setup_bsd_cryptodev();
+#endif
 }

