diff -ur openssl-1.0.2q.orig/Configure openssl-1.0.2q/Configure
--- openssl-1.0.2q.orig/Configure	2015-01-23 00:58:32.000000000 +1000
+++ openssl-1.0.2q/Configure	2015-03-04 13:48:04.264916581 +1000
@@ -680,8 +680,8 @@
 "tandem-c89","c89:-Ww -D__TANDEM -D_XOPEN_SOURCE -D_XOPEN_SOURCE_EXTENDED=1 -D_TANDEM_SOURCE -DB_ENDIAN::(unknown):::THIRTY_TWO_BIT:::",
 
 # uClinux
-"uClinux-dist","$ENV{'CC'}:\$(CFLAGS)::-D_REENTRANT::\$(LDFLAGS) \$(LDLIBS):BN_LLONG:${no_asm}:$ENV{'LIBSSL_dlfcn'}:linux-shared:-fPIC:-shared:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR):$ENV{'RANLIB'}::",
-"uClinux-dist64","$ENV{'CC'}:\$(CFLAGS)::-D_REENTRANT::\$(LDFLAGS) \$(LDLIBS):SIXTY_FOUR_BIT_LONG:${no_asm}:$ENV{'LIBSSL_dlfcn'}:linux-shared:-fPIC:-shared:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR):$ENV{'RANLIB'}::",
+"uClinux-dist","$ENV{'CC'}:\$(CFLAGS)::-D_REENTRANT::\$(LDFLAGS) \$(LDLIBS):BN_LLONG:".(eval "\$$ENV{'UCDIST_ASM'}").":$ENV{'LIBSSL_dlfcn'}:linux-shared:-fPIC:-shared:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR):$ENV{'RANLIB'}::",
+"uClinux-dist64","$ENV{'CC'}:\$(CFLAGS)::-D_REENTRANT::\$(LDFLAGS) \$(LDLIBS):SIXTY_FOUR_BIT_LONG:".(eval "\$$ENV{'UCDIST_ASM'}").":\$(UCDIST_TYP):$ENV{'LIBSSL_dlfcn'}:linux-shared:-fPIC:-shared:.so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR):$ENV{'RANLIB'}::",
 
 );
 
diff -ur openssl-1.0.2q.orig/crypto/arm_arch.h openssl-1.0.2q/crypto/arm_arch.h
--- openssl-1.0.2q.orig/crypto/arm_arch.h	2015-01-23 00:58:32.000000000 +1000
+++ openssl-1.0.2q/crypto/arm_arch.h	2015-03-04 13:51:06.976919536 +1000
@@ -42,6 +42,8 @@
 #    define __ARM_ARCH__ 5
 #   elif defined(__ARM_ARCH_4__) || defined(__ARM_ARCH_4T__)
 #    define __ARM_ARCH__ 4
+#   elif defined(__ARM_ARCH_3__) || defined(__ARM_ARCH_3T__)
+#    define __ARM_ARCH__ 3
 #   else
 #    error "unsupported ARM architecture"
 #   endif
--- openssl-1.0.2q.orig/crypto/evp/c_all.c	2018-04-13 14:06:39.955785330 +1000
+++ openssl-1.0.2q/crypto/evp/c_all.c	2018-04-13 14:06:59.035785744 +1000
@@ -82,4 +82,7 @@
     OPENSSL_cpuid_setup();
     OpenSSL_add_all_ciphers();
     OpenSSL_add_all_digests();
+#if defined(OPENSSL_NO_HW) && (defined(__OpenBSD__) || defined(__FreeBSD__) || defined(HAVE_CRYPTODEV))
+    ENGINE_setup_bsd_cryptodev();
+#endif
 }

--- openssl-1.0.2q.orig/crypto/engine/eng_cryptodev.c	2019-01-03 10:45:39.503532619 +1000
+++ openssl-1.0.2q/crypto/engine/eng_cryptodev.c	2019-01-03 16:05:15.183493704 +1000
@@ -199,8 +199,16 @@
         CRYPTO_SHA1_HMAC, NID_hmacWithSHA1, 20
     },
     {
-        CRYPTO_RIPEMD160_HMAC, NID_ripemd160, 16
-        /* ? */
+        CRYPTO_SHA2_256_HMAC, NID_hmacWithSHA256, 32
+    },
+    {
+        CRYPTO_SHA2_384_HMAC, NID_hmacWithSHA384, 48
+    },
+    {
+        CRYPTO_SHA2_512_HMAC, NID_hmacWithSHA512, 64
+    },
+    {
+        CRYPTO_RIPEMD160_HMAC, NID_ripemd160, 20
     },
     {
         CRYPTO_MD5_KPDK, NID_undef, 0
@@ -215,6 +223,15 @@
         CRYPTO_SHA1, NID_sha1, 20
     },
     {
+        CRYPTO_SHA2_256, NID_sha256, 32
+    },
+    {
+        CRYPTO_SHA2_384, NID_sha384, 48
+    },
+    {
+        CRYPTO_SHA2_512, NID_sha512, 64
+    },
+    {
         0, NID_undef, 0
     },
 };
@@ -342,8 +359,9 @@
         sess.mackeylen = digests[i].keylen;
         sess.cipher = 0;
         if (ioctl(fd, CIOCGSESSION, &sess) != -1 &&
-            ioctl(fd, CIOCFSESSION, &sess.ses) != -1)
+            ioctl(fd, CIOCFSESSION, &sess.ses) != -1) {
             nids[count++] = digests[i].nid;
+	}
     }
     put_dev_crypto(fd);
 
@@ -966,6 +984,51 @@
     sizeof(struct dev_crypto_state),
 };
 
+const EVP_MD cryptodev_sha256 = {
+    NID_sha256,
+    NID_undef,
+    SHA256_DIGEST_LENGTH,
+    EVP_MD_FLAG_ONESHOT,
+    cryptodev_digest_init,
+    cryptodev_digest_update,
+    cryptodev_digest_final,
+    cryptodev_digest_copy,
+    cryptodev_digest_cleanup,
+    EVP_PKEY_NULL_method,
+    SHA256_CBLOCK,
+    sizeof(struct dev_crypto_state),
+};
+
+const EVP_MD cryptodev_sha384 = {
+    NID_sha384,
+    NID_undef,
+    SHA384_DIGEST_LENGTH,
+    EVP_MD_FLAG_ONESHOT,
+    cryptodev_digest_init,
+    cryptodev_digest_update,
+    cryptodev_digest_final,
+    cryptodev_digest_copy,
+    cryptodev_digest_cleanup,
+    EVP_PKEY_NULL_method,
+    SHA512_CBLOCK,
+    sizeof(struct dev_crypto_state),
+};
+
+const EVP_MD cryptodev_sha512 = {
+    NID_sha512,
+    NID_undef,
+    SHA512_DIGEST_LENGTH,
+    EVP_MD_FLAG_ONESHOT,
+    cryptodev_digest_init,
+    cryptodev_digest_update,
+    cryptodev_digest_final,
+    cryptodev_digest_copy,
+    cryptodev_digest_cleanup,
+    EVP_PKEY_NULL_method,
+    SHA512_CBLOCK,
+    sizeof(struct dev_crypto_state),
+};
+
 const EVP_MD cryptodev_md5 = {
     NID_md5,
     NID_undef,
@@ -998,6 +1061,15 @@
     case NID_sha1:
         *digest = &cryptodev_sha1;
         break;
+    case NID_sha256:
+        *digest = &cryptodev_sha256;
+        break;
+    case NID_sha384:
+        *digest = &cryptodev_sha384;
+        break;
+    case NID_sha512:
+        *digest = &cryptodev_sha512;
+        break;
     default:
 # endif                         /* USE_CRYPTODEV_DIGESTS */
         *digest = NULL;
