
Allow algvs tests to be built with no-asm

--- openssl-fips-2.0.10/test/fips_algvs.c.orig	2016-08-18 14:40:30.851318827 +1000
+++ openssl-fips-2.0.10/test/fips_algvs.c	2016-08-18 14:40:32.519318854 +1000
@@ -326,7 +326,7 @@
 	SysInit();
 #endif
 
-#if (defined(__arm__) || defined(__aarch64__))
+#if !defined(OPENSSL_NO_ASM) && (defined(__arm__) || defined(__aarch64__))
 	if (*args && !strcmp(*args, "-noaccel"))
 		{
 		extern unsigned int OPENSSL_armcap_P;
--- openssl-fips-2.0.10/fips/fips_test_suite.c.orig	2016-08-18 15:01:16.843338979 +1000
+++ openssl-fips-2.0.10/fips/fips_test_suite.c	2016-08-18 15:01:18.935339013 +1000
@@ -1325,7 +1325,7 @@
 
     FIPS_post_set_callback(post_cb);
 
-#if (defined(__arm__) || defined(__aarch64__))
+#if !defined(OPENSSL_NO_ASM) && (defined(__arm__) || defined(__aarch64__))
     extern unsigned int OPENSSL_armcap_P;
     if (0 == OPENSSL_armcap_P)
 	fprintf(stderr, "Optimizations disabled\n");
