Index: pam_radius-release_2_0_0/src/pam_radius_auth.c
===================================================================
--- pam_radius-release_2_0_0.orig/src/pam_radius_auth.c
+++ pam_radius-release_2_0_0/src/pam_radius_auth.c
@@ -1242,7 +1242,6 @@ PAM_EXTERN int pam_sm_authenticate(pam_h
 
 	if (password) {
 		password = strdup(password);
-		DPRINT(LOG_DEBUG, "Got password %s", password);
 	}
 
 	/* no previous password: maybe get one from the user */
