--- conntrack-tools-1.4.2/include/fds.h.org	2018-09-10 22:58:45.242161416 +1000
+++ conntrack-tools-1.4.2/include/fds.h	2018-09-10 22:58:24.968301119 +1000
@@ -1,6 +1,7 @@
 #ifndef _FDS_H_
 #define _FDS_H_
 
+#include <sys/select.h>
 #include "linux_list.h"
 
 struct fds {
--- conntrack-tools-1.4.2/include/mcast.h.org	2018-09-10 23:00:28.436715154 +1000
+++ conntrack-tools-1.4.2/include/mcast.h	2018-09-10 23:00:38.484356139 +1000
@@ -1,6 +1,7 @@
 #ifndef _MCAST_H_
 #define _MCAST_H_
 
+#include <sys/select.h>
 #include <stdint.h>
 #include <netinet/in.h>
 #include <net/if.h>
