--- src/pppoe.h.org	2019-06-06 23:49:28.200371130 +1000
+++ src/pppoe.h	2019-06-06 23:49:45.227679667 +1000
@@ -127,10 +127,6 @@
 #error Could not find a 32-bit integer type
 #endif
 
-#ifdef HAVE_LINUX_IF_ETHER_H
-#include <linux/if_ether.h>
-#endif
-
 #include <netinet/in.h>
 
 #ifdef HAVE_NETINET_IF_ETHER_H
@@ -144,6 +140,9 @@
 #endif
 #endif
 
+#ifdef HAVE_LINUX_IF_ETHER_H
+#include <linux/if_ether.h>
+#endif
 
 
 /* Ethernet frame types according to RFC 2516 */
