Index: valgrind-3.21.0/configure
===================================================================
--- valgrind-3.21.0.orig/configure
+++ valgrind-3.21.0/configure
@@ -6873,7 +6873,7 @@ printf "%s\n" "ok (${host_cpu})" >&6; }
         ARCH_MAX="s390x"
         ;;
 
-     armv7*)
+     arm*)
 	{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: ok (${host_cpu})" >&5
 printf "%s\n" "ok (${host_cpu})" >&6; }
 	ARCH_MAX="arm"
